// @google/adk/dist/cjs/tools/vertex_rag_retrieval_tool.js
/**
  * @license
  * Copyright 2026 Google LLC
  * SPDX-License-Identifier: Apache-2.0
  */

"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var vertex_rag_retrieval_tool_exports = {};
__export(vertex_rag_retrieval_tool_exports, {
  VertexRagRetrievalTool: () => VertexRagRetrievalTool
});
module.exports = __toCommonJS(vertex_rag_retrieval_tool_exports);
var import_base_tool = require("./base_tool.js");
/**
 * @license
 * Copyright 2026 Google LLC
 * SPDX-License-Identifier: Apache-2.0
 */
class VertexRagRetrievalTool extends import_base_tool.BaseTool {
  constructor(config) {
    super({
      name: "vertex_rag_retrieval",
      description: "Vertex AI RAG Retrieval Tool"
    });
    this.vertexRagStore = config;
  }
  /**
   * This tool is executed server-side by the Vertex AI RAG Engine.
   * Local execution is not required.
   */
  runAsync() {
    return Promise.resolve();
  }
  async processLlmRequest({
    llmRequest
  }) {
    llmRequest.config = llmRequest.config || {};
    llmRequest.config.tools = llmRequest.config.tools || [];
    llmRequest.config.tools.push({
      retrieval: { vertexRagStore: this.vertexRagStore }
    });
  }
}
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  VertexRagRetrievalTool
});
