// @google/adk/dist/cjs/integrations/agent_registry/agent_registry_mcp_toolset.js
/**
  * @license
  * Copyright 2026 Google LLC
  * SPDX-License-Identifier: Apache-2.0
  */

"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var agent_registry_mcp_toolset_exports = {};
__export(agent_registry_mcp_toolset_exports, {
  AgentRegistrySingleMCPToolset: () => AgentRegistrySingleMCPToolset
});
module.exports = __toCommonJS(agent_registry_mcp_toolset_exports);
var import_base_toolset = require("../../tools/base_toolset.js");
var import_mcp_session_manager = require("../../tools/mcp/mcp_session_manager.js");
var import_mcp_tool = require("../../tools/mcp/mcp_tool.js");
var import_types2 = require("./types.js");
/**
 * @license
 * Copyright 2026 Google LLC
 * SPDX-License-Identifier: Apache-2.0
 */
class AgentRegistrySingleMCPToolset extends import_base_toolset.BaseToolset {
  /**
   * @param options - Configuration for the MCP toolset.
   * @param options.destinationResourceId - Telemetry identifier injected as
   *   `gcp.mcp.server.destination.id` into each resolved tool's custom metadata.
   * @param options.connectionParams - HTTP connection parameters for the MCP server.
   * @param options.toolFilter - Optional predicate or list of tool names to include.
   *   When omitted, all tools from the server are returned.
   * @param options.prefix - Optional prefix prepended to each tool name (e.g. `myServer_toolName`).
   * @param options.headerProvider - Optional async function called immediately before each
   *   {@link getTools} invocation to supply or refresh request headers (e.g. GCP auth tokens).
   * @param options.authScheme - Optional auth scheme forwarded to each resolved tool.
   * @param options.authCredential - Optional credential forwarded to each resolved tool.
   */
  constructor(options) {
    super(options.toolFilter || [], options.prefix);
    this.destinationResourceId = options.destinationResourceId;
    this.connectionParams = options.connectionParams;
    this.headerProvider = options.headerProvider;
    this.authScheme = options.authScheme;
    this.authCredential = options.authCredential;
  }
  /**
   * Connects to the underlying MCP server, retrieves tool definitions, prefixes
   * tool names, and injects destination telemetry metadata into each tool.
   *
   * The `headerProvider`, if configured, is invoked immediately before the
   * connection is established so that tokens are always fresh.
   *
   * @param context - Optional readonly agent context passed to the header provider.
   * @returns The resolved and optionally filtered list of {@link MCPTool} instances.
   */
  async getTools(context) {
    var _a, _b, _c;
    const headers = {};
    if (this.headerProvider) {
      const providerHeaders = await this.headerProvider(context);
      Object.assign(headers, providerHeaders);
    }
    const connectionParamsCopy = {
      ...this.connectionParams,
      transportOptions: {
        ...this.connectionParams.transportOptions,
        requestInit: {
          ...(_a = this.connectionParams.transportOptions) == null ? void 0 : _a.requestInit,
          headers: {
            ...(_c = (_b = this.connectionParams.transportOptions) == null ? void 0 : _b.requestInit) == null ? void 0 : _c.headers,
            ...headers
          }
        }
      }
    };
    const sessionManager = new import_mcp_session_manager.MCPSessionManager(connectionParamsCopy);
    const session = await sessionManager.createSession();
    const listResult = await session.listTools();
    const tools = listResult.tools.map((tool) => {
      const prefixedName = this.prefix ? `${this.prefix}_${tool.name}` : tool.name;
      const mcpTool = new import_mcp_tool.MCPTool(
        { ...tool, name: prefixedName },
        sessionManager,
        tool.name
      );
      const toolWithMetadata = mcpTool;
      if (this.destinationResourceId) {
        if (!toolWithMetadata.customMetadata) {
          toolWithMetadata.customMetadata = {};
        }
        toolWithMetadata.customMetadata[import_types2.GCP_MCP_SERVER_DESTINATION_ID] = this.destinationResourceId;
      }
      return mcpTool;
    });
    const filter = this.toolFilter;
    if (!filter || Array.isArray(filter) && filter.length === 0) {
      return tools;
    }
    return tools.filter((t) => this.isToolSelected(t, context));
  }
  async close() {
  }
}
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  AgentRegistrySingleMCPToolset
});
