// @google/adk/dist/cjs/integrations/agent_registry/types.js
/**
  * @license
  * Copyright 2026 Google LLC
  * SPDX-License-Identifier: Apache-2.0
  */

"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var types_exports = {};
__export(types_exports, {
  AGENT_REGISTRY_BASE_URL: () => AGENT_REGISTRY_BASE_URL,
  GCP_MCP_SERVER_DESTINATION_ID: () => GCP_MCP_SERVER_DESTINATION_ID,
  ProtocolType: () => ProtocolType
});
module.exports = __toCommonJS(types_exports);
/**
 * @license
 * Copyright 2026 Google LLC
 * SPDX-License-Identifier: Apache-2.0
 */
const AGENT_REGISTRY_BASE_URL = "https://agentregistry.googleapis.com/v1alpha";
const GCP_MCP_SERVER_DESTINATION_ID = "gcp.mcp.server.destination.id";
var ProtocolType = /* @__PURE__ */ ((ProtocolType2) => {
  ProtocolType2["TYPE_UNSPECIFIED"] = "TYPE_UNSPECIFIED";
  ProtocolType2["A2A_AGENT"] = "A2A_AGENT";
  ProtocolType2["CUSTOM"] = "CUSTOM";
  return ProtocolType2;
})(ProtocolType || {});
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  AGENT_REGISTRY_BASE_URL,
  GCP_MCP_SERVER_DESTINATION_ID,
  ProtocolType
});
