// @google/adk/dist/esm/integrations/agent_registry/agent_registry_mcp_toolset.js
import {createRequire as topLevelCreateRequire} from 'module';
const require = topLevelCreateRequire(import.meta.url);
/**
 * @license
 * Copyright 2026 Google LLC
 * SPDX-License-Identifier: Apache-2.0
 */
import { BaseToolset } from "../../tools/base_toolset.js";
import {
  MCPSessionManager
} from "../../tools/mcp/mcp_session_manager.js";
import { MCPTool } from "../../tools/mcp/mcp_tool.js";
import { GCP_MCP_SERVER_DESTINATION_ID } from "./types.js";
class AgentRegistrySingleMCPToolset extends BaseToolset {
  /**
   * @param options - Configuration for the MCP toolset.
   * @param options.destinationResourceId - Telemetry identifier injected as
   *   `gcp.mcp.server.destination.id` into each resolved tool's custom metadata.
   * @param options.connectionParams - HTTP connection parameters for the MCP server.
   * @param options.toolFilter - Optional predicate or list of tool names to include.
   *   When omitted, all tools from the server are returned.
   * @param options.prefix - Optional prefix prepended to each tool name (e.g. `myServer_toolName`).
   * @param options.headerProvider - Optional async function called immediately before each
   *   {@link getTools} invocation to supply or refresh request headers (e.g. GCP auth tokens).
   * @param options.authScheme - Optional auth scheme forwarded to each resolved tool.
   * @param options.authCredential - Optional credential forwarded to each resolved tool.
   */
  constructor(options) {
    super(options.toolFilter || [], options.prefix);
    this.destinationResourceId = options.destinationResourceId;
    this.connectionParams = options.connectionParams;
    this.headerProvider = options.headerProvider;
    this.authScheme = options.authScheme;
    this.authCredential = options.authCredential;
  }
  /**
   * Connects to the underlying MCP server, retrieves tool definitions, prefixes
   * tool names, and injects destination telemetry metadata into each tool.
   *
   * The `headerProvider`, if configured, is invoked immediately before the
   * connection is established so that tokens are always fresh.
   *
   * @param context - Optional readonly agent context passed to the header provider.
   * @returns The resolved and optionally filtered list of {@link MCPTool} instances.
   */
  async getTools(context) {
    var _a, _b, _c;
    const headers = {};
    if (this.headerProvider) {
      const providerHeaders = await this.headerProvider(context);
      Object.assign(headers, providerHeaders);
    }
    const connectionParamsCopy = {
      ...this.connectionParams,
      transportOptions: {
        ...this.connectionParams.transportOptions,
        requestInit: {
          ...(_a = this.connectionParams.transportOptions) == null ? void 0 : _a.requestInit,
          headers: {
            ...(_c = (_b = this.connectionParams.transportOptions) == null ? void 0 : _b.requestInit) == null ? void 0 : _c.headers,
            ...headers
          }
        }
      }
    };
    const sessionManager = new MCPSessionManager(connectionParamsCopy);
    const session = await sessionManager.createSession();
    const listResult = await session.listTools();
    const tools = listResult.tools.map((tool) => {
      const prefixedName = this.prefix ? `${this.prefix}_${tool.name}` : tool.name;
      const mcpTool = new MCPTool(
        { ...tool, name: prefixedName },
        sessionManager,
        tool.name
      );
      const toolWithMetadata = mcpTool;
      if (this.destinationResourceId) {
        if (!toolWithMetadata.customMetadata) {
          toolWithMetadata.customMetadata = {};
        }
        toolWithMetadata.customMetadata[GCP_MCP_SERVER_DESTINATION_ID] = this.destinationResourceId;
      }
      return mcpTool;
    });
    const filter = this.toolFilter;
    if (!filter || Array.isArray(filter) && filter.length === 0) {
      return tools;
    }
    return tools.filter((t) => this.isToolSelected(t, context));
  }
  async close() {
  }
}
export {
  AgentRegistrySingleMCPToolset
};
