// @google/adk/dist/cjs/tools/openapi_tool/openapi_spec_parser/openapi_spec_parser.js
/**
  * @license
  * Copyright 2026 Google LLC
  * SPDX-License-Identifier: Apache-2.0
  */

"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var __decorateClass = (decorators, target, key, kind) => {
  var result = kind > 1 ? void 0 : kind ? __getOwnPropDesc(target, key) : target;
  for (var i = decorators.length - 1, decorator; i >= 0; i--)
    if (decorator = decorators[i])
      result = (kind ? decorator(target, key, result) : decorator(result)) || result;
  if (kind && result) __defProp(target, key, result);
  return result;
};
var openapi_spec_parser_exports = {};
__export(openapi_spec_parser_exports, {
  OpenApiSpecParser: () => OpenApiSpecParser
});
module.exports = __toCommonJS(openapi_spec_parser_exports);
var import_experimental = require("../../../utils/experimental.js");
var import_operation_parser = require("./operation_parser.js");
/**
 * @license
 * Copyright 2026 Google LLC
 * SPDX-License-Identifier: Apache-2.0
 */
const VALID_SCHEMA_TYPES = /* @__PURE__ */ new Set([
  "array",
  "boolean",
  "integer",
  "null",
  "number",
  "object",
  "string"
]);
let OpenApiSpecParser = class {
  constructor(options = {}) {
    var _a;
    this.preservePropertyNames = (_a = options.preservePropertyNames) != null ? _a : false;
  }
  parse(openapiSpec) {
    const resolvedSpec = resolveReferences(openapiSpec);
    const sanitizedSpec = sanitizeSchemaTypes(resolvedSpec);
    return collectOperations(sanitizedSpec, {
      preservePropertyNames: this.preservePropertyNames
    });
  }
};
__decorateClass([
  import_experimental.experimental
], OpenApiSpecParser.prototype, "parse", 1);
OpenApiSpecParser = __decorateClass([
  import_experimental.experimental
], OpenApiSpecParser);
function resolveReferences(spec) {
  const resolvedCache = /* @__PURE__ */ new Map();
  const specCopy = JSON.parse(JSON.stringify(spec));
  const recursiveResolve = (obj, currentDoc, seenRefs = /* @__PURE__ */ new Set()) => {
    if (typeof obj !== "object" || obj === null) {
      return obj;
    }
    if (Array.isArray(obj)) {
      return obj.map((item) => recursiveResolve(item, currentDoc, seenRefs));
    }
    const objRecord = obj;
    if ("$ref" in objRecord && typeof objRecord["$ref"] === "string") {
      const refString = objRecord["$ref"];
      if (seenRefs.has(refString) && !resolvedCache.has(refString)) {
        const copy = { ...objRecord };
        delete copy["$ref"];
        return copy;
      }
      seenRefs.add(refString);
      if (resolvedCache.has(refString)) {
        return resolvedCache.get(refString);
      }
      let resolvedValue = resolveRef(refString, currentDoc);
      if (resolvedValue !== void 0) {
        resolvedValue = recursiveResolve(resolvedValue, currentDoc, seenRefs);
        resolvedCache.set(refString, resolvedValue);
        return resolvedValue;
      } else {
        return obj;
      }
    }
    const newDict = {};
    for (const [key, value] of Object.entries(objRecord)) {
      newDict[key] = recursiveResolve(value, currentDoc, seenRefs);
    }
    return newDict;
  };
  return recursiveResolve(specCopy, specCopy);
}
function resolveRef(refString, currentDoc) {
  const parts = refString.split("/");
  if (parts[0] !== "#") {
    throw new Error(`External references not supported: ${refString}`);
  }
  let current = currentDoc;
  for (const part of parts.slice(1)) {
    if (typeof current === "object" && current !== null && part in current) {
      current = current[part];
    } else {
      return void 0;
    }
  }
  return current;
}
function sanitizeSchemaTypes(openapiSpec) {
  const specCopy = JSON.parse(JSON.stringify(openapiSpec));
  const sanitizeTypeField = (schemaDict) => {
    if (!("type" in schemaDict)) return;
    const typeValue = schemaDict["type"];
    if (typeof typeValue === "string") {
      const normalizedType = typeValue.toLowerCase();
      if (VALID_SCHEMA_TYPES.has(normalizedType)) {
        schemaDict["type"] = normalizedType;
      } else {
        delete schemaDict["type"];
      }
      return;
    }
    if (Array.isArray(typeValue)) {
      const validTypes = [];
      for (const entry of typeValue) {
        if (typeof entry !== "string") continue;
        const normalizedEntry = entry.toLowerCase();
        if (VALID_SCHEMA_TYPES.has(normalizedEntry) && !validTypes.includes(normalizedEntry)) {
          validTypes.push(normalizedEntry);
        }
      }
      if (validTypes.length > 0) {
        schemaDict["type"] = validTypes;
      } else {
        delete schemaDict["type"];
      }
    }
  };
  const sanitizeRecursive = (obj, inSchema) => {
    if (typeof obj !== "object" || obj === null) {
      return obj;
    }
    if (Array.isArray(obj)) {
      return obj.map((item) => sanitizeRecursive(item, inSchema));
    }
    const objRecord = obj;
    if (inSchema) {
      sanitizeTypeField(objRecord);
    }
    for (const [key, value] of Object.entries(objRecord)) {
      const isSchemaContainer = key === "schema" || key === "schemas";
      objRecord[key] = sanitizeRecursive(value, inSchema || isSchemaContainer);
    }
    return objRecord;
  };
  return sanitizeRecursive(specCopy, false);
}
function collectOperations(spec, options = {}) {
  var _a, _b, _c, _d;
  const preservePropertyNames = (_a = options.preservePropertyNames) != null ? _a : false;
  const operations = [];
  const baseUrl = ((_c = (_b = spec.servers) == null ? void 0 : _b[0]) == null ? void 0 : _c.url) || "";
  const globalSecurity = spec.security || [];
  let globalSchemeName;
  if (globalSecurity.length > 0) {
    globalSchemeName = Object.keys(globalSecurity[0])[0];
  }
  const authSchemes = ((_d = spec.components) == null ? void 0 : _d.securitySchemes) || {};
  for (const [path, pathItem] of Object.entries(spec.paths || {})) {
    if (!pathItem) continue;
    const methods = [
      "get",
      "post",
      "put",
      "delete",
      "patch",
      "head",
      "options",
      "trace"
    ];
    for (const method of methods) {
      const operation = pathItem[method];
      if (!operation) continue;
      const pathParams = pathItem.parameters || [];
      const opParams = operation.parameters || [];
      operation.parameters = [...opParams, ...pathParams];
      if (!operation.operationId) {
        operation.operationId = `${method}_${path.replace(/[^a-zA-Z0-9]/g, "_")}`;
      }
      const parser = new import_operation_parser.OperationParser(operation, {
        preservePropertyNames
      });
      let authSchemeName;
      if (operation.security && operation.security.length > 0) {
        authSchemeName = Object.keys(operation.security[0])[0];
      }
      authSchemeName = authSchemeName || globalSchemeName;
      const authScheme = authSchemeName ? authSchemes[authSchemeName] : void 0;
      operations.push({
        name: parser.getFunctionName(),
        description: parser.getDescription(),
        endpoint: { baseUrl, path, method },
        operation,
        parameters: parser.getParameters(),
        authScheme
      });
    }
  }
  return operations;
}
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  OpenApiSpecParser
});
