// @google/adk/dist/cjs/tools/openapi_tool/openapi_spec_parser/operation_parser.js
/**
  * @license
  * Copyright 2026 Google LLC
  * SPDX-License-Identifier: Apache-2.0
  */

"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var __decorateClass = (decorators, target, key, kind) => {
  var result = kind > 1 ? void 0 : kind ? __getOwnPropDesc(target, key) : target;
  for (var i = decorators.length - 1, decorator; i >= 0; i--)
    if (decorator = decorators[i])
      result = (kind ? decorator(target, key, result) : decorator(result)) || result;
  if (kind && result) __defProp(target, key, result);
  return result;
};
var operation_parser_exports = {};
__export(operation_parser_exports, {
  OperationParser: () => OperationParser
});
module.exports = __toCommonJS(operation_parser_exports);
var import_experimental = require("../../../utils/experimental.js");
/**
 * @license
 * Copyright 2026 Google LLC
 * SPDX-License-Identifier: Apache-2.0
 */
let OperationParser = class {
  constructor(operation, options = {}) {
    this.operation = operation;
    this.params = [];
    var _a;
    this.preservePropertyNames = (_a = options.preservePropertyNames) != null ? _a : false;
    this.processOperationParameters();
    this.processRequestBody();
    this.processReturnValue();
    this.dedupeParamNames();
  }
  getParamName(originalName) {
    if (this.preservePropertyNames) {
      return originalName;
    }
    return originalName.replace(/[A-Z]/g, (g) => "_" + g.toLowerCase()).replace(/^_/, "");
  }
  processOperationParameters() {
    const parameters = this.operation.parameters || [];
    for (const param of parameters) {
      if ("name" in param) {
        const originalName = param.name;
        const description = param.description || "";
        const location = param.in || "";
        const schema = param.schema || {};
        this.params.push({
          originalName,
          paramLocation: location,
          paramSchema: schema,
          description,
          required: param.required || false,
          name: this.getParamName(originalName)
        });
      }
    }
  }
  processRequestBody() {
    const requestBody = this.operation.requestBody;
    if (!requestBody || "$ref" in requestBody) {
      return;
    }
    const content = requestBody.content || {};
    const firstMimeType = Object.keys(content)[0];
    if (!firstMimeType) {
      return;
    }
    const mediaTypeObject = content[firstMimeType];
    const schema = mediaTypeObject.schema;
    const description = requestBody.description || "";
    if (schema && !("$ref" in schema)) {
      if (schema.type === "object") {
        const properties = schema.properties || {};
        if (Object.keys(properties).length > 0) {
          for (const [propName, propDetails] of Object.entries(properties)) {
            if (!("$ref" in propDetails)) {
              this.params.push({
                originalName: propName,
                paramLocation: "body",
                paramSchema: propDetails,
                description: propDetails.description,
                required: (schema.required || []).includes(propName),
                name: this.getParamName(propName)
              });
            }
          }
        } else {
          this.params.push({
            originalName: "",
            paramLocation: "body",
            paramSchema: schema,
            description,
            required: true,
            name: "body"
          });
        }
      } else if (schema.type === "array") {
        this.params.push({
          originalName: "array",
          paramLocation: "body",
          paramSchema: schema,
          description,
          required: true,
          name: "body"
        });
      } else {
        this.params.push({
          originalName: "body",
          paramLocation: "body",
          paramSchema: schema,
          description,
          required: true,
          name: "body"
        });
      }
    }
  }
  processReturnValue() {
    const responses = this.operation.responses || {};
    const validCodes = Object.keys(responses).filter((k) => k.startsWith("2"));
    const min20x = validCodes.sort()[0];
    let returnSchema = {};
    if (min20x) {
      const response = responses[min20x];
      if (!("$ref" in response) && response.content) {
        const firstMimeType = Object.keys(response.content)[0];
        if (firstMimeType) {
          const schema = response.content[firstMimeType].schema;
          if (schema && !("$ref" in schema)) {
            returnSchema = schema;
          }
        }
      }
    }
    this.returnValue = {
      originalName: "",
      paramLocation: "",
      paramSchema: returnSchema,
      required: true,
      name: "return"
    };
  }
  dedupeParamNames() {
    const nameCounts = /* @__PURE__ */ new Map();
    for (const param of this.params) {
      const name = param.name;
      const count = nameCounts.get(name) || 0;
      if (count > 0) {
        param.name = `${name}_${count}`;
      }
      nameCounts.set(name, count + 1);
    }
  }
  getParameters() {
    return this.params;
  }
  getJsonSchema() {
    const properties = {};
    const required = [];
    for (const param of this.params) {
      properties[param.name] = param.paramSchema;
      if (param.required) {
        required.push(param.name);
      }
    }
    return {
      type: "object",
      properties,
      required: required.length > 0 ? required : void 0,
      title: `${this.operation.operationId || "unnamed"}_Arguments`
    };
  }
  getFunctionName() {
    const operationId = this.operation.operationId;
    if (!operationId) {
      throw new Error("Operation ID is missing");
    }
    return this.getParamName(operationId).substring(0, 60);
  }
  getDescription() {
    return this.operation.description || this.operation.summary || "";
  }
};
__decorateClass([
  import_experimental.experimental
], OperationParser.prototype, "getParameters", 1);
__decorateClass([
  import_experimental.experimental
], OperationParser.prototype, "getJsonSchema", 1);
__decorateClass([
  import_experimental.experimental
], OperationParser.prototype, "getFunctionName", 1);
__decorateClass([
  import_experimental.experimental
], OperationParser.prototype, "getDescription", 1);
OperationParser = __decorateClass([
  import_experimental.experimental
], OperationParser);
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  OperationParser
});
