// @google/adk/dist/cjs/tools/openapi_tool/openapi_spec_parser/tool_auth_handler.js
/**
  * @license
  * Copyright 2026 Google LLC
  * SPDX-License-Identifier: Apache-2.0
  */

"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var __decorateClass = (decorators, target, key, kind) => {
  var result = kind > 1 ? void 0 : kind ? __getOwnPropDesc(target, key) : target;
  for (var i = decorators.length - 1, decorator; i >= 0; i--)
    if (decorator = decorators[i])
      result = (kind ? decorator(target, key, result) : decorator(result)) || result;
  if (kind && result) __defProp(target, key, result);
  return result;
};
var tool_auth_handler_exports = {};
__export(tool_auth_handler_exports, {
  ToolAuthHandler: () => ToolAuthHandler
});
module.exports = __toCommonJS(tool_auth_handler_exports);
var import_experimental = require("../../../utils/experimental.js");
var import_auto_auth_credential_exchanger = require("../auth/credential_exchangers/auto_auth_credential_exchanger.js");
/**
 * @license
 * Copyright 2026 Google LLC
 * SPDX-License-Identifier: Apache-2.0
 */
class ToolContextCredentialStore {
  constructor(context) {
    this.context = context;
  }
  getCredentialKey(authScheme) {
    const schemeName = (authScheme == null ? void 0 : authScheme.type) || "default";
    return `${schemeName}_existing_exchanged_credential`;
  }
  getCredential(authScheme) {
    const key = this.getCredentialKey(authScheme);
    const state = this.context.state;
    if (state) {
      const serialized = state[key];
      if (serialized) {
        return serialized;
      }
    }
    return void 0;
  }
  storeCredential(key, credential) {
    const state = this.context.state;
    if (state) {
      state[key] = credential;
    }
  }
}
let ToolAuthHandler = class {
  constructor(context, authScheme, authCredential, credentialKey) {
    this.context = context;
    this.authScheme = authScheme;
    this.authCredential = authCredential;
    this.credentialKey = credentialKey;
  }
  static fromToolContext(context, authScheme, authCredential, options = {}) {
    return new ToolAuthHandler(
      context,
      authScheme,
      authCredential,
      options.credentialKey
    );
  }
  async prepareAuthCredentials() {
    if (!this.authScheme) {
      return { state: "done" };
    }
    const store = new ToolContextCredentialStore(this.context);
    const existingCredential = store.getCredential(this.authScheme);
    if (existingCredential) {
      return { state: "done", authCredential: existingCredential };
    }
    const authConfig = {
      authScheme: this.authScheme,
      rawAuthCredential: this.authCredential,
      credentialKey: this.credentialKey || "default_openapi_key"
    };
    const credential = this.context.getAuthResponse(authConfig);
    if (credential) {
      const exchanger = new import_auto_auth_credential_exchanger.AutoAuthCredentialExchanger();
      const result = await exchanger.exchange({
        authScheme: this.authScheme,
        authCredential: credential
      });
      const key = store.getCredentialKey(this.authScheme);
      store.storeCredential(key, result.credential);
      return { state: "done", authCredential: result.credential };
    }
    this.context.requestCredential(authConfig);
    return { state: "pending" };
  }
};
__decorateClass([
  import_experimental.experimental
], ToolAuthHandler.prototype, "prepareAuthCredentials", 1);
__decorateClass([
  import_experimental.experimental
], ToolAuthHandler, "fromToolContext", 1);
ToolAuthHandler = __decorateClass([
  import_experimental.experimental
], ToolAuthHandler);
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  ToolAuthHandler
});
