// @google/adk/dist/web/tools/openapi_tool/openapi_toolset.js
import {createRequire as topLevelCreateRequire} from 'module';
const require = topLevelCreateRequire(import.meta.url);
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __decorateClass = (decorators, target, key, kind) => {
  var result = kind > 1 ? void 0 : kind ? __getOwnPropDesc(target, key) : target;
  for (var i = decorators.length - 1, decorator; i >= 0; i--)
    if (decorator = decorators[i])
      result = (kind ? decorator(target, key, result) : decorator(result)) || result;
  if (kind && result) __defProp(target, key, result);
  return result;
};
/**
 * @license
 * Copyright 2026 Google LLC
 * SPDX-License-Identifier: Apache-2.0
 */
import yaml from "js-yaml";
import { experimental } from "../../utils/experimental.js";
import { BaseToolset } from "../base_toolset.js";
import { OpenApiSpecParser } from "./openapi_spec_parser/openapi_spec_parser.js";
import { createRestApiTool } from "./rest_api_tool.js";
let OpenAPIToolset = class extends BaseToolset {
  constructor(options = {}) {
    super(options.toolFilter || [], options.prefix);
    this.tools = [];
    let spec = options.specDict;
    if (!spec && options.specStr) {
      if (options.specType === "yaml" || !options.specType && options.specStr.trim().startsWith("---")) {
        spec = yaml.load(options.specStr);
      } else {
        spec = JSON.parse(options.specStr);
      }
    }
    if (!spec) {
      throw new Error("Either specDict or specStr must be provided.");
    }
    const parser = new OpenApiSpecParser({
      preservePropertyNames: options.preservePropertyNames
    });
    const parsedOperations = parser.parse(spec);
    for (const op of parsedOperations) {
      let toolName = op.name;
      if (this.prefix) {
        toolName = "".concat(this.prefix, "_").concat(toolName);
      }
      const tool = createRestApiTool(
        {
          name: toolName,
          description: op.description,
          endpoint: op.endpoint,
          operation: op.operation,
          authScheme: op.authScheme
        },
        {
          preservePropertyNames: options.preservePropertyNames,
          headerProvider: options.headerProvider,
          credentialKey: options.credentialKey
        }
      );
      this.tools.push(tool);
    }
    if (options.authScheme || options.authCredential) {
      for (const tool of this.tools) {
        if (options.authScheme) tool.configureAuthScheme(options.authScheme);
        if (options.authCredential)
          tool.configureAuthCredential(options.authCredential);
      }
    }
  }
  async getTools(context) {
    return this.tools.filter((tool) => {
      if (Array.isArray(this.toolFilter) && this.toolFilter.length > 0) {
        return this.toolFilter.includes(tool.name);
      }
      if (context) {
        return this.isToolSelected(tool, context);
      }
      return true;
    });
  }
  async close() {
    return Promise.resolve();
  }
};
__decorateClass([
  experimental
], OpenAPIToolset.prototype, "getTools", 1);
__decorateClass([
  experimental
], OpenAPIToolset.prototype, "close", 1);
OpenAPIToolset = __decorateClass([
  experimental
], OpenAPIToolset);
export {
  OpenAPIToolset
};
