// @google/adk/dist/cjs/tools/openapi_tool/openapi_toolset.js
/**
  * @license
  * Copyright 2026 Google LLC
  * SPDX-License-Identifier: Apache-2.0
  */

"use strict";
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var __decorateClass = (decorators, target, key, kind) => {
  var result = kind > 1 ? void 0 : kind ? __getOwnPropDesc(target, key) : target;
  for (var i = decorators.length - 1, decorator; i >= 0; i--)
    if (decorator = decorators[i])
      result = (kind ? decorator(target, key, result) : decorator(result)) || result;
  if (kind && result) __defProp(target, key, result);
  return result;
};
var openapi_toolset_exports = {};
__export(openapi_toolset_exports, {
  OpenAPIToolset: () => OpenAPIToolset
});
module.exports = __toCommonJS(openapi_toolset_exports);
var import_js_yaml = __toESM(require("js-yaml"), 1);
var import_experimental = require("../../utils/experimental.js");
var import_base_toolset = require("../base_toolset.js");
var import_openapi_spec_parser = require("./openapi_spec_parser/openapi_spec_parser.js");
var import_rest_api_tool = require("./rest_api_tool.js");
/**
 * @license
 * Copyright 2026 Google LLC
 * SPDX-License-Identifier: Apache-2.0
 */
let OpenAPIToolset = class extends import_base_toolset.BaseToolset {
  constructor(options = {}) {
    super(options.toolFilter || [], options.prefix);
    this.tools = [];
    let spec = options.specDict;
    if (!spec && options.specStr) {
      if (options.specType === "yaml" || !options.specType && options.specStr.trim().startsWith("---")) {
        spec = import_js_yaml.default.load(options.specStr);
      } else {
        spec = JSON.parse(options.specStr);
      }
    }
    if (!spec) {
      throw new Error("Either specDict or specStr must be provided.");
    }
    const parser = new import_openapi_spec_parser.OpenApiSpecParser({
      preservePropertyNames: options.preservePropertyNames
    });
    const parsedOperations = parser.parse(spec);
    for (const op of parsedOperations) {
      let toolName = op.name;
      if (this.prefix) {
        toolName = `${this.prefix}_${toolName}`;
      }
      const tool = (0, import_rest_api_tool.createRestApiTool)(
        {
          name: toolName,
          description: op.description,
          endpoint: op.endpoint,
          operation: op.operation,
          authScheme: op.authScheme
        },
        {
          preservePropertyNames: options.preservePropertyNames,
          headerProvider: options.headerProvider,
          credentialKey: options.credentialKey
        }
      );
      this.tools.push(tool);
    }
    if (options.authScheme || options.authCredential) {
      for (const tool of this.tools) {
        if (options.authScheme) tool.configureAuthScheme(options.authScheme);
        if (options.authCredential)
          tool.configureAuthCredential(options.authCredential);
      }
    }
  }
  async getTools(context) {
    return this.tools.filter((tool) => {
      if (Array.isArray(this.toolFilter) && this.toolFilter.length > 0) {
        return this.toolFilter.includes(tool.name);
      }
      if (context) {
        return this.isToolSelected(tool, context);
      }
      return true;
    });
  }
  async close() {
    return Promise.resolve();
  }
};
__decorateClass([
  import_experimental.experimental
], OpenAPIToolset.prototype, "getTools", 1);
__decorateClass([
  import_experimental.experimental
], OpenAPIToolset.prototype, "close", 1);
OpenAPIToolset = __decorateClass([
  import_experimental.experimental
], OpenAPIToolset);
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  OpenAPIToolset
});
