// @google/adk/dist/cjs/tools/openapi_tool/rest_api_tool.js
/**
  * @license
  * Copyright 2026 Google LLC
  * SPDX-License-Identifier: Apache-2.0
  */

"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var __decorateClass = (decorators, target, key, kind) => {
  var result = kind > 1 ? void 0 : kind ? __getOwnPropDesc(target, key) : target;
  for (var i = decorators.length - 1, decorator; i >= 0; i--)
    if (decorator = decorators[i])
      result = (kind ? decorator(target, key, result) : decorator(result)) || result;
  if (kind && result) __defProp(target, key, result);
  return result;
};
var rest_api_tool_exports = {};
__export(rest_api_tool_exports, {
  RestApiTool: () => RestApiTool,
  createRestApiTool: () => createRestApiTool,
  prepareRequestBody: () => prepareRequestBody,
  prepareRequestParams: () => prepareRequestParams
});
module.exports = __toCommonJS(rest_api_tool_exports);
var import_experimental = require("../../utils/experimental.js");
var import_base_tool = require("../base_tool.js");
var import_auth_helpers = require("./auth/auth_helpers.js");
var import_operation_parser = require("./openapi_spec_parser/operation_parser.js");
var import_tool_auth_handler = require("./openapi_spec_parser/tool_auth_handler.js");
/**
 * @license
 * Copyright 2026 Google LLC
 * SPDX-License-Identifier: Apache-2.0
 */
let RestApiTool = class extends import_base_tool.BaseTool {
  constructor(name, description, endpoint, operation, authScheme, authCredential, options = {}) {
    super({ name, description });
    this.endpoint = endpoint;
    this.operation = operation;
    this.authScheme = authScheme;
    this.authCredential = authCredential;
    this.authScheme = authScheme;
    this.authCredential = authCredential;
    this.headerProvider = options.headerProvider;
    this.credentialKey = options.credentialKey;
    this.operationParser = new import_operation_parser.OperationParser(operation, options);
  }
  configureAuthScheme(authScheme) {
    this.authScheme = authScheme;
  }
  configureAuthCredential(authCredential) {
    this.authCredential = authCredential;
  }
  configureCredentialKey(credentialKey) {
    this.credentialKey = credentialKey;
  }
  _getDeclaration() {
    const schema = this.operationParser.getJsonSchema();
    return {
      name: this.name,
      description: this.description,
      parameters: schema
    };
  }
  async runAsync(request) {
    const context = request.toolContext;
    const args = request.args;
    const authHandler = import_tool_auth_handler.ToolAuthHandler.fromToolContext(
      context,
      this.authScheme,
      this.authCredential,
      { credentialKey: this.credentialKey }
    );
    const authResult = await authHandler.prepareAuthCredentials();
    if (authResult.state === "pending") {
      return {
        pending: true,
        message: "Needs your authorization to access your data."
      };
    }
    const credential = authResult.authCredential;
    const method = this.endpoint.method.toUpperCase();
    const {
      url: initialUrl,
      headers,
      body: parsedBody,
      bodyData
    } = prepareRequestParams(
      this.endpoint,
      this.operationParser.getParameters(),
      args
    );
    const body = prepareRequestBody(
      this.operation.requestBody,
      parsedBody,
      bodyData,
      headers
    );
    const url = (0, import_auth_helpers.applyCredential)(
      initialUrl,
      headers,
      credential,
      this.authScheme
    );
    if (this.headerProvider) {
      const providerHeaders = this.headerProvider(context);
      Object.assign(headers, providerHeaders);
    }
    try {
      const response = await globalThis.fetch(url, {
        method,
        headers,
        // eslint-disable-next-line no-undef
        body
      });
      const contentType = response.headers.get("content-type");
      if (contentType && contentType.includes("application/json")) {
        return await response.json();
      } else {
        return await response.text();
      }
    } catch (error) {
      return {
        error: `Failed to execute API call: ${error.message}`
      };
    }
  }
};
__decorateClass([
  import_experimental.experimental
], RestApiTool.prototype, "configureAuthScheme", 1);
__decorateClass([
  import_experimental.experimental
], RestApiTool.prototype, "configureAuthCredential", 1);
__decorateClass([
  import_experimental.experimental
], RestApiTool.prototype, "configureCredentialKey", 1);
__decorateClass([
  import_experimental.experimental
], RestApiTool.prototype, "_getDeclaration", 1);
__decorateClass([
  import_experimental.experimental
], RestApiTool.prototype, "runAsync", 1);
RestApiTool = __decorateClass([
  import_experimental.experimental
], RestApiTool);
function prepareRequestParams(endpoint, parameters, args) {
  const headers = {};
  const queryParams = new URLSearchParams();
  let body = void 0;
  const paramsMap = new Map(parameters.map((p) => [p.name, p]));
  const pathParams = {};
  const bodyData = {};
  for (const [argName, argValue] of Object.entries(args)) {
    const param = paramsMap.get(argName);
    if (!param) continue;
    const originalName = param.originalName;
    const location = param.paramLocation;
    if (location === "path") {
      pathParams[originalName] = String(argValue);
    } else if (location === "query") {
      queryParams.append(originalName, String(argValue));
    } else if (location === "header") {
      headers[originalName] = String(argValue);
    } else if (location === "body") {
      if (originalName === "body" || originalName === "array" || originalName === "") {
        body = argValue;
      } else {
        bodyData[originalName] = argValue;
      }
    }
  }
  let url = `${endpoint.baseUrl}${endpoint.path}`;
  for (const [key, value] of Object.entries(pathParams)) {
    url = url.replace(`{${key}}`, value);
  }
  const urlParts = url.split("?");
  if (urlParts.length > 1) {
    const pathQueryParams = new URLSearchParams(urlParts[1]);
    for (const [key, value] of pathQueryParams.entries()) {
      queryParams.append(key, value);
    }
    url = urlParts[0];
  }
  const queryString = queryParams.toString();
  if (queryString) {
    url += `?${queryString}`;
  }
  return { url, headers, body, bodyData };
}
function prepareRequestBody(requestBody, body, bodyData, headers) {
  const finalData = body !== void 0 ? body : Object.keys(bodyData).length > 0 ? bodyData : void 0;
  if (requestBody && "content" in requestBody) {
    const content = requestBody.content;
    for (const [mimeType, _mediaTypeObject] of Object.entries(content)) {
      if (finalData !== void 0) {
        if (mimeType === "application/json" || mimeType.endsWith("+json")) {
          headers["Content-Type"] = mimeType;
          return typeof finalData === "string" ? finalData : JSON.stringify(finalData);
        } else if (mimeType === "application/x-www-form-urlencoded") {
          return new URLSearchParams(finalData);
        } else if (mimeType === "multipart/form-data") {
          const formData = new FormData();
          if (typeof finalData === "object" && finalData !== null) {
            for (const [key, value] of Object.entries(finalData)) {
              formData.append(key, String(value));
            }
          }
          return formData;
        } else if (mimeType === "text/plain") {
          headers["Content-Type"] = mimeType;
          return String(finalData);
        }
      }
      break;
    }
  } else if (finalData !== void 0) {
    headers["Content-Type"] = "application/json";
    return typeof finalData === "string" ? finalData : JSON.stringify(finalData);
  }
  return void 0;
}
function createRestApiTool(parsed, options = {}) {
  return new RestApiTool(
    parsed.name,
    parsed.description,
    parsed.endpoint,
    parsed.operation,
    parsed.authScheme,
    void 0,
    options
  );
}
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  RestApiTool,
  createRestApiTool,
  prepareRequestBody,
  prepareRequestParams
});
