// @google/adk/dist/web/tools/openapi_tool/auth/credential_exchangers/service_account_exchanger.js
import {createRequire as topLevelCreateRequire} from 'module';
const require = topLevelCreateRequire(import.meta.url);
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __decorateClass = (decorators, target, key, kind) => {
  var result = kind > 1 ? void 0 : kind ? __getOwnPropDesc(target, key) : target;
  for (var i = decorators.length - 1, decorator; i >= 0; i--)
    if (decorator = decorators[i])
      result = (kind ? decorator(target, key, result) : decorator(result)) || result;
  if (kind && result) __defProp(target, key, result);
  return result;
};
/**
 * @license
 * Copyright 2026 Google LLC
 * SPDX-License-Identifier: Apache-2.0
 */
import { GoogleAuth, JWT } from "google-auth-library";
import {
  AuthCredentialTypes
} from "../../../../auth/auth_credential.js";
import {
  CredentialExchangeError
} from "../../../../auth/exchanger/base_credential_exchanger.js";
import { experimental } from "../../../../utils/experimental.js";
const DEFAULT_SCOPES = ["https://www.googleapis.com/auth/cloud-platform"];
let ServiceAccountCredentialExchanger = class {
  async exchange(params) {
    const { authCredential } = params;
    if (authCredential.authType !== AuthCredentialTypes.SERVICE_ACCOUNT || !authCredential.serviceAccount) {
      throw new CredentialExchangeError(
        "Invalid credential type for ServiceAccountCredentialExchanger"
      );
    }
    const saConfig = authCredential.serviceAccount;
    if (saConfig.useDefaultCredential) {
      return this.exchangeForDefaultCredential(saConfig);
    }
    return this.exchangeForExplicitCredential(saConfig);
  }
  async exchangeForDefaultCredential(saConfig) {
    try {
      const auth = new GoogleAuth({
        scopes: saConfig.scopes || DEFAULT_SCOPES
      });
      const client = await auth.getClient();
      const tokenResponse = await client.getAccessToken();
      const token = tokenResponse.token;
      if (!token) {
        throw new Error("Failed to get access token from default credentials");
      }
      return {
        credential: {
          authType: AuthCredentialTypes.HTTP,
          http: {
            scheme: "bearer",
            credentials: { token }
          }
        },
        wasExchanged: true
      };
    } catch (error) {
      throw new CredentialExchangeError(
        "Failed to exchange default service account token: ".concat(error.message)
      );
    }
  }
  async exchangeForExplicitCredential(saConfig) {
    const creds = saConfig.serviceAccountCredential;
    if (!creds) {
      throw new CredentialExchangeError(
        "Service account credentials are missing."
      );
    }
    try {
      const client = new JWT({
        email: creds.clientEmail,
        key: creds.privateKey,
        scopes: saConfig.scopes
      });
      const tokens = await client.authorize();
      const token = tokens.access_token;
      if (!token) {
        throw new Error("Failed to get access token from explicit credentials");
      }
      return {
        credential: {
          authType: AuthCredentialTypes.HTTP,
          http: {
            scheme: "bearer",
            credentials: { token }
          }
        },
        wasExchanged: true
      };
    } catch (error) {
      throw new CredentialExchangeError(
        "Failed to exchange explicit service account token: ".concat(error.message)
      );
    }
  }
};
__decorateClass([
  experimental
], ServiceAccountCredentialExchanger.prototype, "exchange", 1);
ServiceAccountCredentialExchanger = __decorateClass([
  experimental
], ServiceAccountCredentialExchanger);
export {
  ServiceAccountCredentialExchanger
};
