// @google/adk/dist/web/tools/exit_loop_tool.js
import {createRequire as topLevelCreateRequire} from 'module';
const require = topLevelCreateRequire(import.meta.url);
/**
 * @license
 * Copyright 2026 Google LLC
 * SPDX-License-Identifier: Apache-2.0
 */
import { BaseTool } from "./base_tool.js";
class ExitLoopTool extends BaseTool {
  /** Creates an ExitLoopTool with its fixed name and description. */
  constructor() {
    super({
      name: "exit_loop",
      description: "Exits the loop.\n\nCall this function only when you are instructed to do so."
    });
  }
  /** Returns the function declaration for this tool. */
  _getDeclaration() {
    return {
      name: this.name,
      description: this.description
    };
  }
  /**
   * Sets `escalate` and `skipSummarization` on the event actions, signalling
   * the {@link LoopAgent} to stop iterating.
   *
   * @param request The tool request containing the current tool context.
   * @returns An empty string response.
   */
  async runAsync({
    toolContext
  }) {
    toolContext.actions.escalate = true;
    toolContext.actions.skipSummarization = true;
    return "";
  }
}
const EXIT_LOOP = new ExitLoopTool();
export {
  EXIT_LOOP,
  ExitLoopTool
};
