// @google/adk/dist/cjs/code_executors/agent_engine_sandbox_code_executor.js
/**
  * @license
  * Copyright 2026 Google LLC
  * SPDX-License-Identifier: Apache-2.0
  */

"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var __decorateClass = (decorators, target, key, kind) => {
  var result = kind > 1 ? void 0 : kind ? __getOwnPropDesc(target, key) : target;
  for (var i = decorators.length - 1, decorator; i >= 0; i--)
    if (decorator = decorators[i])
      result = (kind ? decorator(target, key, result) : decorator(result)) || result;
  if (kind && result) __defProp(target, key, result);
  return result;
};
var agent_engine_sandbox_code_executor_exports = {};
__export(agent_engine_sandbox_code_executor_exports, {
  AgentEngineSandboxCodeExecutor: () => AgentEngineSandboxCodeExecutor
});
module.exports = __toCommonJS(agent_engine_sandbox_code_executor_exports);
var import_vertexai = require("@google-cloud/vertexai");
var import_types = require("@google-cloud/vertexai/build/src/genai/types.js");
var import_experimental = require("../utils/experimental.js");
var import_file_utils = require("../utils/file_utils.js");
var import_logger = require("../utils/logger.js");
var import_base_code_executor = require("./base_code_executor.js");
var import_code_execution_utils = require("./code_execution_utils.js");
/**
 * @license
 * Copyright 2026 Google LLC
 * SPDX-License-Identifier: Apache-2.0
 */
const SANDBOX_PATTERN = /^projects\/([a-zA-Z0-9-_]+)\/locations\/([a-zA-Z0-9-_]+)\/reasoningEngines\/(\d+)\/sandboxEnvironments\/(\d+)$/;
const ENGINE_PATTERN = /^projects\/([a-zA-Z0-9-_]+)\/locations\/([a-zA-Z0-9-_]+)\/reasoningEngines\/(\d+)$/;
const DEFAULT_MAX_ATTEMPTS = 180;
const DEFAULT_SANDBOX_TTL = "31536000s";
const DEFAULT_SANDBOX_DISPLAY_NAME = "default_sandbox";
const DEFAULT_ENGINE_DISPLAY_NAME = "default_engine";
function mapLanguage(lang) {
  switch (lang) {
    case import_code_execution_utils.CodeExecutionLanguage.PYTHON:
      return import_types.Language.LANGUAGE_PYTHON;
    case import_code_execution_utils.CodeExecutionLanguage.JAVASCRIPT:
      return import_types.Language.LANGUAGE_JAVASCRIPT;
    default:
      throw new Error(`Unsupported language for Agent Engine Sandbox: ${lang}`);
  }
}
let AgentEngineSandboxCodeExecutor = class extends import_base_code_executor.BaseCodeExecutor {
  constructor(options = {}) {
    super();
    this.sandboxResourceName = options.sandboxResourceName;
    this.agentEngineResourceName = options.agentEngineResourceName;
    this.projectId = options.projectId || process.env.GOOGLE_CLOUD_PROJECT;
    this.location = options.location || process.env.GOOGLE_CLOUD_LOCATION || "us-central1";
    if (this.sandboxResourceName) {
      const match = this.sandboxResourceName.match(SANDBOX_PATTERN);
      if (match) {
        this.projectId = match[1];
        this.location = match[2];
      } else {
        throw new Error(
          `Invalid sandbox resource name: ${this.sandboxResourceName}`
        );
      }
    } else if (this.agentEngineResourceName) {
      const match = this.agentEngineResourceName.match(ENGINE_PATTERN);
      if (match) {
        this.projectId = match[1];
        this.location = match[2];
      } else {
        throw new Error(
          `Invalid agent engine resource name: ${this.agentEngineResourceName}`
        );
      }
    }
    if (options.client) {
      this.client = options.client;
    } else {
      if (!this.projectId) {
        throw new Error("Project ID is required.");
      }
      this.client = new import_vertexai.Client({
        project: this.projectId,
        location: this.location
      });
    }
  }
  async executeCode(params) {
    var _a;
    const { invocationContext, codeExecutionInput } = params;
    const language = mapLanguage(codeExecutionInput.language);
    const agentEngineName = await this.getOrCreateAgentEngine();
    const sandboxName = await this.getOrCreateSandbox(
      invocationContext,
      agentEngineName,
      language
    );
    const inputs = [
      {
        mimeType: "application/json",
        data: Buffer.from(
          JSON.stringify({ code: codeExecutionInput.code })
        ).toString("base64")
      }
    ];
    if (codeExecutionInput.inputFiles) {
      for (const file of codeExecutionInput.inputFiles) {
        inputs.push({
          mimeType: file.mimeType,
          data: file.content,
          // Assumed to be already base64 encoded based on CodeExecutionInput definition
          metadata: {
            attributes: {
              file_name: Buffer.from(file.name).toString("base64")
            }
          }
        });
      }
    }
    import_logger.logger.debug(`Executing code in sandbox ${sandboxName}...`);
    const response = await this.client.agentEnginesInternal.sandboxes.executeCodeInternal({
      name: sandboxName,
      inputs
    });
    let stdout = "";
    let stderr = "";
    const outputFiles = [];
    if (response.outputs) {
      for (const output of response.outputs) {
        const attributes = ((_a = output.metadata) == null ? void 0 : _a.attributes) || {};
        const encodedFileName = attributes["file_name"];
        const fileName = encodedFileName ? Buffer.from(encodedFileName, "base64").toString("utf-8") : void 0;
        if (output.mimeType === "application/json" && !fileName) {
          if (output.data) {
            const jsonStr = Buffer.from(output.data, "base64").toString(
              "utf-8"
            );
            try {
              const jsonData = JSON.parse(jsonStr);
              if (jsonData.msg_out !== void 0) {
                stdout = jsonData.msg_out;
              } else {
                stdout = "";
              }
              if (jsonData.msg_err !== void 0) {
                stderr = jsonData.msg_err;
              } else {
                stderr = "";
              }
            } catch (e) {
              import_logger.logger.warn("Failed to parse JSON output from sandbox", e);
              stdout = jsonStr;
            }
          }
        } else {
          let mimeType = output.mimeType;
          const name = fileName || "output_file";
          if (!mimeType) {
            mimeType = (0, import_file_utils.guessMimeType)(name);
          }
          outputFiles.push({
            name,
            content: output.data || "",
            mimeType
          });
        }
      }
    }
    return {
      stdout,
      stderr,
      outputFiles
    };
  }
  async getOrCreateAgentEngine() {
    if (this.agentEngineResourceName) {
      return this.agentEngineResourceName;
    }
    if (!this.agentEngineCreationPromise) {
      this.agentEngineCreationPromise = (async () => {
        import_logger.logger.debug(
          "No Agent Engine resource name provided. Creating a new one..."
        );
        const operation = await this.client.agentEnginesInternal.createInternal(
          {
            config: {
              displayName: DEFAULT_ENGINE_DISPLAY_NAME
            }
          }
        );
        let apiResponse = operation;
        let attempts = 0;
        while (!apiResponse.done && attempts < DEFAULT_MAX_ATTEMPTS) {
          await new Promise((resolve) => setTimeout(resolve, 1e3));
          apiResponse = await this.client.agentEnginesInternal.getAgentOperationInternal({
            operationName: operation.name
          });
          attempts++;
        }
        if (!apiResponse.done) {
          throw new Error(
            `Agent Engine creation operation ${operation.name} did not complete in time.`
          );
        }
        const response = apiResponse.response;
        this.agentEngineResourceName = response.name;
        import_logger.logger.debug(`Created Agent Engine: ${this.agentEngineResourceName}`);
        return this.agentEngineResourceName;
      })();
    }
    return this.agentEngineCreationPromise;
  }
  async getOrCreateSandbox(invocationContext, agentEngineName, language) {
    var _a, _b;
    if (this.sandboxResourceName) {
      return this.sandboxResourceName;
    }
    const stateKey = `sandbox_name_${language.toLowerCase()}`;
    let sandboxName = (_b = (_a = invocationContext.session) == null ? void 0 : _a.state) == null ? void 0 : _b[stateKey];
    let createNewSandbox = false;
    if (!sandboxName) {
      createNewSandbox = true;
    } else {
      try {
        const sandbox = await this.client.agentEnginesInternal.sandboxes.getInternal({
          name: sandboxName
        });
        if (!sandbox || sandbox.state !== "STATE_RUNNING") {
          createNewSandbox = true;
        }
      } catch (error) {
        import_logger.logger.debug(
          `Failed to get sandbox ${sandboxName}, will create a new one`,
          error
        );
        createNewSandbox = true;
      }
    }
    if (createNewSandbox) {
      import_logger.logger.debug(`Creating a new ${language} code execution sandbox...`);
      const operation = await this.client.agentEnginesInternal.sandboxes.createInternal({
        name: agentEngineName,
        spec: {
          codeExecutionEnvironment: {
            codeLanguage: language
          }
        },
        config: {
          displayName: DEFAULT_SANDBOX_DISPLAY_NAME,
          ttl: DEFAULT_SANDBOX_TTL
        }
      });
      let apiResponse = operation;
      let attempts = 0;
      while (!apiResponse.done && attempts < DEFAULT_MAX_ATTEMPTS) {
        await new Promise((resolve) => setTimeout(resolve, 1e3));
        apiResponse = await this.client.agentEnginesInternal.sandboxes.getSandboxOperationInternal(
          {
            operationName: operation.name
          }
        );
        attempts++;
      }
      if (!apiResponse.done) {
        throw new Error(
          `Sandbox creation operation ${operation.name} did not complete in time.`
        );
      }
      const response = apiResponse.response;
      sandboxName = response.name;
      if (invocationContext.session) {
        if (!invocationContext.session.state) {
          invocationContext.session.state = {};
        }
        invocationContext.session.state[stateKey] = sandboxName;
      }
    }
    return sandboxName;
  }
};
AgentEngineSandboxCodeExecutor = __decorateClass([
  import_experimental.experimental
], AgentEngineSandboxCodeExecutor);
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  AgentEngineSandboxCodeExecutor
});
