// @google/adk/dist/web/integrations/agent_registry/agent_registry.js
import {createRequire as topLevelCreateRequire} from 'module';
const require = topLevelCreateRequire(import.meta.url);
/**
 * @license
 * Copyright 2026 Google LLC
 * SPDX-License-Identifier: Apache-2.0
 */
import { GoogleAuth } from "google-auth-library";
import { RemoteA2AAgent } from "../../a2a/a2a_remote_agent.js";
import { logger } from "../../utils/logger.js";
import { AgentRegistrySingleMCPToolset } from "./agent_registry_mcp_toolset.js";
import { cleanName, isGoogleApi } from "./helpers.js";
import {
  AGENT_REGISTRY_BASE_URL,
  ProtocolType
} from "./types.js";
export * from "./agent_registry_mcp_toolset.js";
export * from "./helpers.js";
export * from "./types.js";
const TRANSPORT_MAPPING = {
  "HTTP_JSON": "HTTP+JSON",
  "JSONRPC": "JSONRPC",
  "GRPC": "GRPC"
};
class AgentRegistry {
  constructor(options) {
    if (!options.projectId || !options.location) {
      throw new Error("project_id and location must be provided");
    }
    this.projectId = options.projectId;
    this.location = options.location;
    this.basePath = "projects/".concat(this.projectId, "/locations/").concat(this.location);
    this.headerProvider = options.headerProvider;
    this.auth = new GoogleAuth({
      scopes: ["https://www.googleapis.com/auth/cloud-platform"]
    });
  }
  /**
   * Resolves default Google Cloud credentials and returns standard headers.
   * Automatically caches, fetches, and handles refreshing expired OAuth tokens.
   * Injects the billing/quota project identifier `x-goog-user-project` if present.
   */
  async getAuthHeaders() {
    try {
      const client = await this.auth.getClient();
      const headers = await client.getRequestHeaders(
        "https://agentregistry.googleapis.com"
      );
      const authHeaders = {};
      const rawHeaders = headers;
      const authKey = Object.keys(rawHeaders).find(
        (k) => k.toLowerCase() === "authorization"
      );
      let token = authKey ? rawHeaders[authKey] : void 0;
      if (!token && client.credentials && client.credentials.access_token) {
        token = "Bearer ".concat(client.credentials.access_token);
      }
      if (token) {
        authHeaders["Authorization"] = token;
      }
      authHeaders["Content-Type"] = "application/json";
      const quotaProjectId = client.quotaProjectId || this.auth.quotaProjectId;
      if (quotaProjectId) {
        authHeaders["x-goog-user-project"] = quotaProjectId;
      }
      return authHeaders;
    } catch (err) {
      const msg = err instanceof Error ? err.message : String(err);
      throw new Error("Failed to refresh Google Cloud credentials: ".concat(msg));
    }
  }
  /**
   * Helper function to execute HTTP GET requests against the Agent Registry API.
   * Handles path resolution, search query params compilation, and auth headers fetching.
   */
  async makeRequest(path, params) {
    let url;
    if (path.startsWith("projects/")) {
      url = "".concat(AGENT_REGISTRY_BASE_URL, "/").concat(path);
    } else {
      url = "".concat(AGENT_REGISTRY_BASE_URL, "/").concat(this.basePath, "/").concat(path);
    }
    if (params && Object.keys(params).length > 0) {
      const searchParams = new URLSearchParams(params);
      url += "?".concat(searchParams.toString());
    }
    try {
      const headers = await this.getAuthHeaders();
      const res = await fetch(url, {
        method: "GET",
        headers
      });
      if (!res.ok) {
        const text = await res.text();
        throw new Error(
          "API request failed with status ".concat(res.status, ": ").concat(text)
        );
      }
      return await res.json();
    } catch (err) {
      const msg = err instanceof Error ? err.message : String(err);
      if (msg.includes("API request failed")) {
        throw err;
      }
      throw new Error("API request failed: ".concat(msg));
    }
  }
  /**
   * Parses connection interfaces list from registry metadata and returns the first match
   * corresponding to requested protocol types and binding options.
   */
  getConnectionUri(resourceDetails, filters) {
    const protocols = [];
    if (resourceDetails.protocols) {
      protocols.push(...resourceDetails.protocols);
    }
    if (resourceDetails.interfaces) {
      protocols.push({ interfaces: resourceDetails.interfaces });
    }
    for (const p of protocols) {
      if ((filters == null ? void 0 : filters.protocolType) && p.type !== filters.protocolType) {
        continue;
      }
      const protocolVersion = p.protocolVersion;
      const interfaces = p.interfaces || [];
      for (const i of interfaces) {
        const mappedBinding = i.protocolBinding ? TRANSPORT_MAPPING[i.protocolBinding] : void 0;
        if ((filters == null ? void 0 : filters.protocolBinding) && mappedBinding !== filters.protocolBinding) {
          continue;
        }
        if (i.url) {
          return { url: i.url, protocolVersion, protocolBinding: mappedBinding };
        }
      }
    }
    return {};
  }
  // --- MCP Server Methods ---
  async listMcpServers(options) {
    const params = {};
    if (options == null ? void 0 : options.filterStr) {
      params["filter"] = options.filterStr;
    }
    if (options == null ? void 0 : options.pageSize) {
      params["pageSize"] = String(options.pageSize);
    }
    if (options == null ? void 0 : options.pageToken) {
      params["pageToken"] = options.pageToken;
    }
    return this.makeRequest("mcpServers", params);
  }
  async getMcpServer(name) {
    return this.makeRequest(name);
  }
  async getMcpToolset(mcpServerName, options) {
    var _a, _b;
    const serverDetails = await this.getMcpServer(mcpServerName);
    const name = cleanName(serverDetails.displayName || mcpServerName);
    const mcpServerId = serverDetails.mcpServerId;
    let endpointUri = this.getConnectionUri(serverDetails, {
      protocolBinding: "JSONRPC"
    }).url;
    if (!endpointUri) {
      endpointUri = this.getConnectionUri(serverDetails, {
        protocolBinding: "HTTP+JSON"
      }).url;
    }
    if (!endpointUri) {
      throw new Error(
        "MCP Server endpoint URI not found for: ".concat(mcpServerName)
      );
    }
    let authScheme = options == null ? void 0 : options.authScheme;
    if (mcpServerId && !authScheme) {
      try {
        const bindingsData = await this.makeRequest("bindings");
        const bindings = bindingsData.bindings || [];
        for (const b of bindings) {
          const targetId = ((_a = b.target) == null ? void 0 : _a.identifier) || "";
          if (targetId.endsWith(mcpServerId)) {
            const authProvider = (_b = b.authProviderBinding) == null ? void 0 : _b.authProvider;
            if (authProvider) {
              authScheme = {
                type: "gcpAuthProviderScheme",
                name: authProvider,
                continueUri: options == null ? void 0 : options.continueUri
              };
              break;
            }
          }
        }
      } catch (err) {
        const msg = err instanceof Error ? err.message : String(err);
        logger.warn(
          "Failed to fetch bindings for MCP Server ".concat(mcpServerName, ": ").concat(msg)
        );
      }
    }
    const connectionParams = {
      type: "StreamableHTTPConnectionParams",
      url: endpointUri
    };
    const combinedHeaderProvider = async (context) => {
      const headers = {};
      if (!authScheme && !(options == null ? void 0 : options.authCredential) && isGoogleApi(endpointUri)) {
        Object.assign(headers, await this.getAuthHeaders());
      }
      if (this.headerProvider && context) {
        Object.assign(headers, this.headerProvider(context));
      }
      return headers;
    };
    return new AgentRegistrySingleMCPToolset({
      destinationResourceId: mcpServerId,
      connectionParams,
      prefix: name,
      headerProvider: combinedHeaderProvider,
      authScheme,
      authCredential: options == null ? void 0 : options.authCredential
    });
  }
  // --- Endpoint Methods ---
  async listEndpoints(options) {
    const params = {};
    if (options == null ? void 0 : options.filterStr) {
      params["filter"] = options.filterStr;
    }
    if (options == null ? void 0 : options.pageSize) {
      params["pageSize"] = String(options.pageSize);
    }
    if (options == null ? void 0 : options.pageToken) {
      params["pageToken"] = options.pageToken;
    }
    return this.makeRequest("endpoints", params);
  }
  async getEndpoint(name) {
    return this.makeRequest(name);
  }
  async getModelName(endpointName) {
    const endpointDetails = await this.getEndpoint(endpointName);
    const { url } = this.getConnectionUri(endpointDetails);
    if (!url) {
      throw new Error("Connection URI not found for endpoint: ".concat(endpointName));
    }
    const uri = url.replace(/:\w+$/, "");
    if (uri.startsWith("projects/")) {
      return uri;
    }
    const match = uri.match(/(projects\/.+)/);
    if (match) {
      return match[1];
    }
    return uri;
  }
  // --- Agent Methods ---
  async listAgents(options) {
    const params = {};
    if (options == null ? void 0 : options.filterStr) {
      params["filter"] = options.filterStr;
    }
    if (options == null ? void 0 : options.pageSize) {
      params["pageSize"] = String(options.pageSize);
    }
    if (options == null ? void 0 : options.pageToken) {
      params["pageToken"] = options.pageToken;
    }
    return this.makeRequest("agents", params);
  }
  async getAgentInfo(name) {
    return this.makeRequest(name);
  }
  async getRemoteA2AAgent(agentName, options) {
    const agentInfo = await this.getAgentInfo(agentName);
    const card = agentInfo.card || {};
    const cardContent = card.content;
    if (card.type === "A2A_AGENT_CARD" && cardContent) {
      const agentCard2 = cardContent;
      const name2 = cleanName(agentCard2.name);
      return new RemoteA2AAgent({
        name: name2,
        agentCard: agentCard2,
        description: agentCard2.description,
        client: options == null ? void 0 : options.client,
        clientFactory: options == null ? void 0 : options.clientFactory
      });
    }
    const name = cleanName(agentInfo.displayName || agentName);
    const description = agentInfo.description || "";
    const version = agentInfo.version || "";
    const { url, protocolVersion, protocolBinding } = this.getConnectionUri(
      agentInfo,
      {
        protocolType: ProtocolType.A2A_AGENT
      }
    );
    if (!url) {
      throw new Error("A2A connection URI not found for Agent: ".concat(agentName));
    }
    const skills = (agentInfo.skills || []).map(
      (s) => ({
        id: s.id,
        name: s.name,
        description: s.description || "",
        tags: s.tags || [],
        examples: s.examples || []
      })
    );
    const agentCard = {
      name,
      description,
      version,
      preferredTransport: protocolBinding || "HTTP+JSON",
      protocolVersion: protocolVersion || "0.3.0",
      url,
      skills,
      capabilities: {
        streaming: false
      },
      defaultInputModes: ["text"],
      defaultOutputModes: ["text"]
    };
    return new RemoteA2AAgent({
      name,
      agentCard,
      description,
      client: options == null ? void 0 : options.client,
      clientFactory: options == null ? void 0 : options.clientFactory
    });
  }
}
export {
  AgentRegistry
};
