// @google/adk/dist/cjs/tools/openapi_tool/auth/credential_exchangers/service_account_exchanger.js
/**
  * @license
  * Copyright 2026 Google LLC
  * SPDX-License-Identifier: Apache-2.0
  */

"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var __decorateClass = (decorators, target, key, kind) => {
  var result = kind > 1 ? void 0 : kind ? __getOwnPropDesc(target, key) : target;
  for (var i = decorators.length - 1, decorator; i >= 0; i--)
    if (decorator = decorators[i])
      result = (kind ? decorator(target, key, result) : decorator(result)) || result;
  if (kind && result) __defProp(target, key, result);
  return result;
};
var service_account_exchanger_exports = {};
__export(service_account_exchanger_exports, {
  ServiceAccountCredentialExchanger: () => ServiceAccountCredentialExchanger
});
module.exports = __toCommonJS(service_account_exchanger_exports);
var import_google_auth_library = require("google-auth-library");
var import_auth_credential = require("../../../../auth/auth_credential.js");
var import_base_credential_exchanger = require("../../../../auth/exchanger/base_credential_exchanger.js");
var import_experimental = require("../../../../utils/experimental.js");
/**
 * @license
 * Copyright 2026 Google LLC
 * SPDX-License-Identifier: Apache-2.0
 */
const DEFAULT_SCOPES = ["https://www.googleapis.com/auth/cloud-platform"];
let ServiceAccountCredentialExchanger = class {
  async exchange(params) {
    const { authCredential } = params;
    if (authCredential.authType !== import_auth_credential.AuthCredentialTypes.SERVICE_ACCOUNT || !authCredential.serviceAccount) {
      throw new import_base_credential_exchanger.CredentialExchangeError(
        "Invalid credential type for ServiceAccountCredentialExchanger"
      );
    }
    const saConfig = authCredential.serviceAccount;
    if (saConfig.useDefaultCredential) {
      return this.exchangeForDefaultCredential(saConfig);
    }
    return this.exchangeForExplicitCredential(saConfig);
  }
  async exchangeForDefaultCredential(saConfig) {
    try {
      const auth = new import_google_auth_library.GoogleAuth({
        scopes: saConfig.scopes || DEFAULT_SCOPES
      });
      const client = await auth.getClient();
      const tokenResponse = await client.getAccessToken();
      const token = tokenResponse.token;
      if (!token) {
        throw new Error("Failed to get access token from default credentials");
      }
      return {
        credential: {
          authType: import_auth_credential.AuthCredentialTypes.HTTP,
          http: {
            scheme: "bearer",
            credentials: { token }
          }
        },
        wasExchanged: true
      };
    } catch (error) {
      throw new import_base_credential_exchanger.CredentialExchangeError(
        `Failed to exchange default service account token: ${error.message}`
      );
    }
  }
  async exchangeForExplicitCredential(saConfig) {
    const creds = saConfig.serviceAccountCredential;
    if (!creds) {
      throw new import_base_credential_exchanger.CredentialExchangeError(
        "Service account credentials are missing."
      );
    }
    try {
      const client = new import_google_auth_library.JWT({
        email: creds.clientEmail,
        key: creds.privateKey,
        scopes: saConfig.scopes
      });
      const tokens = await client.authorize();
      const token = tokens.access_token;
      if (!token) {
        throw new Error("Failed to get access token from explicit credentials");
      }
      return {
        credential: {
          authType: import_auth_credential.AuthCredentialTypes.HTTP,
          http: {
            scheme: "bearer",
            credentials: { token }
          }
        },
        wasExchanged: true
      };
    } catch (error) {
      throw new import_base_credential_exchanger.CredentialExchangeError(
        `Failed to exchange explicit service account token: ${error.message}`
      );
    }
  }
};
__decorateClass([
  import_experimental.experimental
], ServiceAccountCredentialExchanger.prototype, "exchange", 1);
ServiceAccountCredentialExchanger = __decorateClass([
  import_experimental.experimental
], ServiceAccountCredentialExchanger);
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  ServiceAccountCredentialExchanger
});
