// @google/adk/dist/web/tools/openapi_tool/openapi_spec_parser/tool_auth_handler.js
import {createRequire as topLevelCreateRequire} from 'module';
const require = topLevelCreateRequire(import.meta.url);
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __decorateClass = (decorators, target, key, kind) => {
  var result = kind > 1 ? void 0 : kind ? __getOwnPropDesc(target, key) : target;
  for (var i = decorators.length - 1, decorator; i >= 0; i--)
    if (decorator = decorators[i])
      result = (kind ? decorator(target, key, result) : decorator(result)) || result;
  if (kind && result) __defProp(target, key, result);
  return result;
};
/**
 * @license
 * Copyright 2026 Google LLC
 * SPDX-License-Identifier: Apache-2.0
 */
import { experimental } from "../../../utils/experimental.js";
import { AutoAuthCredentialExchanger } from "../auth/credential_exchangers/auto_auth_credential_exchanger.js";
class ToolContextCredentialStore {
  constructor(context) {
    this.context = context;
  }
  getCredentialKey(authScheme) {
    const schemeName = (authScheme == null ? void 0 : authScheme.type) || "default";
    return "".concat(schemeName, "_existing_exchanged_credential");
  }
  getCredential(authScheme) {
    const key = this.getCredentialKey(authScheme);
    const state = this.context.state;
    if (state) {
      const serialized = state[key];
      if (serialized) {
        return serialized;
      }
    }
    return void 0;
  }
  storeCredential(key, credential) {
    const state = this.context.state;
    if (state) {
      state[key] = credential;
    }
  }
}
let ToolAuthHandler = class {
  constructor(context, authScheme, authCredential, credentialKey) {
    this.context = context;
    this.authScheme = authScheme;
    this.authCredential = authCredential;
    this.credentialKey = credentialKey;
  }
  static fromToolContext(context, authScheme, authCredential, options = {}) {
    return new ToolAuthHandler(
      context,
      authScheme,
      authCredential,
      options.credentialKey
    );
  }
  async prepareAuthCredentials() {
    if (!this.authScheme) {
      return { state: "done" };
    }
    const store = new ToolContextCredentialStore(this.context);
    const existingCredential = store.getCredential(this.authScheme);
    if (existingCredential) {
      return { state: "done", authCredential: existingCredential };
    }
    const authConfig = {
      authScheme: this.authScheme,
      rawAuthCredential: this.authCredential,
      credentialKey: this.credentialKey || "default_openapi_key"
    };
    const credential = this.context.getAuthResponse(authConfig);
    if (credential) {
      const exchanger = new AutoAuthCredentialExchanger();
      const result = await exchanger.exchange({
        authScheme: this.authScheme,
        authCredential: credential
      });
      const key = store.getCredentialKey(this.authScheme);
      store.storeCredential(key, result.credential);
      return { state: "done", authCredential: result.credential };
    }
    this.context.requestCredential(authConfig);
    return { state: "pending" };
  }
};
__decorateClass([
  experimental
], ToolAuthHandler.prototype, "prepareAuthCredentials", 1);
__decorateClass([
  experimental
], ToolAuthHandler, "fromToolContext", 1);
ToolAuthHandler = __decorateClass([
  experimental
], ToolAuthHandler);
export {
  ToolAuthHandler
};
