// @google/adk/dist/cjs/utils/async_queue.js
/**
  * @license
  * Copyright 2026 Google LLC
  * SPDX-License-Identifier: Apache-2.0
  */

"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var async_queue_exports = {};
__export(async_queue_exports, {
  AsyncQueue: () => AsyncQueue
});
module.exports = __toCommonJS(async_queue_exports);
/**
 * @license
 * Copyright 2026 Google LLC
 * SPDX-License-Identifier: Apache-2.0
 */
class AsyncQueue {
  constructor() {
    this.queue = [];
    this.resolvers = [];
    this.closed = false;
  }
  push(value) {
    if (this.closed) return;
    if (this.resolvers.length > 0) {
      const { resolve } = this.resolvers.shift();
      resolve({ value, done: false });
    } else {
      this.queue.push(value);
    }
  }
  error(err) {
    this.errorVal = err;
    while (this.resolvers.length > 0) {
      const { reject } = this.resolvers.shift();
      reject(err);
    }
  }
  close() {
    this.closed = true;
    while (this.resolvers.length > 0) {
      const { resolve } = this.resolvers.shift();
      resolve({ value: void 0, done: true });
    }
  }
  [Symbol.asyncIterator]() {
    return {
      next: () => {
        if (this.errorVal) {
          const err = this.errorVal;
          this.errorVal = void 0;
          return Promise.reject(err);
        }
        if (this.queue.length > 0) {
          return Promise.resolve({ value: this.queue.shift(), done: false });
        }
        if (this.closed) {
          return Promise.resolve({ value: void 0, done: true });
        }
        return new Promise((resolve, reject) => {
          this.resolvers.push({ resolve, reject });
        });
      }
    };
  }
}
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  AsyncQueue
});
