// @google/adk/dist/cjs/utils/streaming_utils.js
/**
  * @license
  * Copyright 2026 Google LLC
  * SPDX-License-Identifier: Apache-2.0
  */

"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var streaming_utils_exports = {};
__export(streaming_utils_exports, {
  StreamingResponseAggregator: () => StreamingResponseAggregator,
  isEmptyContentPart: () => isEmptyContentPart
});
module.exports = __toCommonJS(streaming_utils_exports);
var import_genai = require("@google/genai");
var import_jsonpath_plus = require("jsonpath-plus");
var import_functions = require("../agents/functions.js");
var import_feature_registry = require("../features/feature_registry.js");
var import_llm_response = require("../models/llm_response.js");
/**
 * @license
 * Copyright 2026 Google LLC
 * SPDX-License-Identifier: Apache-2.0
 */
const UNSAFE_PROPERTY_KEYS = /* @__PURE__ */ new Set(["__proto__", "constructor", "prototype"]);
class ProgressiveStrategy {
  constructor() {
    this.partsSequence = [];
    this.currentTextBuffer = "";
    this.currentFcArgs = {};
  }
  flushTextBufferToSequence() {
    if (!this.currentTextBuffer) {
      return;
    }
    if (this.currentTextIsThought) {
      this.partsSequence.push({
        text: this.currentTextBuffer,
        thought: true
      });
    } else {
      this.partsSequence.push({
        text: this.currentTextBuffer
      });
    }
    this.currentTextBuffer = "";
    this.currentTextIsThought = void 0;
  }
  getValueFromPartialArg(partialArg, jsonPath) {
    let value = null;
    let hasValue = false;
    const stringValue = partialArg.stringValue;
    const numberValue = partialArg.numberValue;
    const boolValue = partialArg.boolValue;
    const nullValue = partialArg.nullValue;
    if (stringValue !== void 0) {
      const stringChunk = stringValue;
      hasValue = true;
      const pathParts = import_jsonpath_plus.JSONPath.toPathArray(jsonPath).filter(
        (p) => p !== "$" && p !== "$["
      );
      let existingValue = this.currentFcArgs;
      for (const part of pathParts) {
        if (existingValue && typeof existingValue === "object" && part in existingValue) {
          existingValue = existingValue[part];
        } else {
          existingValue = void 0;
          break;
        }
      }
      if (typeof existingValue === "string") {
        value = existingValue + stringChunk;
      } else {
        value = stringChunk;
      }
    } else if (numberValue !== void 0) {
      value = numberValue;
      hasValue = true;
    } else if (boolValue !== void 0) {
      value = boolValue;
      hasValue = true;
    } else if (nullValue !== void 0) {
      value = null;
      hasValue = true;
    }
    return [value, hasValue];
  }
  setValueByJsonPath(jsonPath, value) {
    const pathParts = import_jsonpath_plus.JSONPath.toPathArray(jsonPath).filter(
      (p) => p !== "$" && p !== "$["
    );
    if (pathParts.some((part) => UNSAFE_PROPERTY_KEYS.has(String(part)))) {
      return;
    }
    let current = this.currentFcArgs;
    for (let i = 0; i < pathParts.length - 1; i++) {
      const part = pathParts[i];
      if (!Object.prototype.hasOwnProperty.call(current, part)) {
        current[part] = {};
      }
      current = current[part];
    }
    if (pathParts.length > 0) {
      current[pathParts[pathParts.length - 1]] = value;
    }
  }
  flushFunctionCallToSequence() {
    var _a;
    if (this.currentFcName) {
      const fcPart = {
        functionCall: {
          name: this.currentFcName,
          args: JSON.parse(JSON.stringify(this.currentFcArgs)),
          id: (_a = this.currentFcId) != null ? _a : (0, import_functions.generateClientFunctionCallId)()
        }
      };
      if (this.currentThoughtSignature) {
        fcPart.thoughtSignature = this.currentThoughtSignature.toString();
      }
      this.partsSequence.push(fcPart);
      this.currentFcName = void 0;
      this.currentFcArgs = {};
      this.currentFcId = void 0;
      this.currentThoughtSignature = void 0;
    }
  }
  processStreamingFunctionCall(fc) {
    if (fc.name) {
      this.currentFcName = fc.name;
    }
    if (fc.id) {
      this.currentFcId = fc.id;
    }
    for (const partialArg of fc.partialArgs || []) {
      const jsonPath = partialArg.jsonPath;
      if (!jsonPath) {
        continue;
      }
      const [value, hasValue] = this.getValueFromPartialArg(
        partialArg,
        jsonPath
      );
      if (hasValue) {
        this.setValueByJsonPath(jsonPath, value);
      }
    }
    if (!fc.willContinue) {
      this.flushTextBufferToSequence();
      this.flushFunctionCallToSequence();
    }
  }
  processFunctionCallPart(part) {
    const fc = part.functionCall;
    if (!fc) {
      return;
    }
    if (fc.partialArgs || fc.willContinue) {
      if (!fc.id && !this.currentFcId) {
        fc.id = (0, import_functions.generateClientFunctionCallId)();
      }
      if (part.thoughtSignature && !this.currentThoughtSignature) {
        this.currentThoughtSignature = part.thoughtSignature;
      }
      this.processStreamingFunctionCall(fc);
    } else {
      if (fc.name) {
        if (!fc.id) {
          fc.id = (0, import_functions.generateClientFunctionCallId)();
        }
        this.flushTextBufferToSequence();
        this.partsSequence.push(part);
      }
    }
  }
  async *processResponse(llmResponse) {
    var _a;
    if (llmResponse.content && llmResponse.content.parts) {
      for (const part of llmResponse.content.parts) {
        if (part.text) {
          const isThought = (_a = part.thought) != null ? _a : false;
          if (this.currentTextBuffer && isThought !== this.currentTextIsThought) {
            this.flushTextBufferToSequence();
          }
          if (!this.currentTextBuffer) {
            this.currentTextIsThought = isThought;
          }
          this.currentTextBuffer += part.text;
        } else if (part.functionCall) {
          this.processFunctionCallPart(part);
        } else {
          this.flushTextBufferToSequence();
          this.partsSequence.push(part);
        }
      }
    }
    llmResponse.partial = true;
    yield llmResponse;
  }
  close() {
    this.flushTextBufferToSequence();
    this.flushFunctionCallToSequence();
    const finalParts = this.partsSequence;
    if (finalParts.length === 0) {
      return void 0;
    }
    return finalParts;
  }
}
class NonProgressiveStrategy {
  constructor() {
    this.text = "";
    this.thoughtText = "";
  }
  async *processResponse(llmResponse) {
    var _a;
    if ((_a = llmResponse.content) == null ? void 0 : _a.parts) {
      const nonTextParts = [];
      let sawTextPart = false;
      for (const part of llmResponse.content.parts) {
        if (typeof part.text === "string") {
          sawTextPart = true;
          if (part.thought) {
            this.thoughtText += part.text;
          } else {
            this.text += part.text;
          }
          continue;
        }
        if (part.functionCall && !part.functionCall.id) {
          part.functionCall.id = (0, import_functions.generateClientFunctionCallId)();
        }
        nonTextParts.push(part);
      }
      if (nonTextParts.length > 0) {
        if (this.thoughtText || this.text) {
          const parts = [];
          if (this.thoughtText) {
            parts.push({ text: this.thoughtText, thought: true });
          }
          if (this.text) {
            parts.push({ text: this.text });
          }
          yield {
            content: {
              role: "model",
              parts
            },
            usageMetadata: llmResponse.usageMetadata,
            partial: false
          };
          this.thoughtText = "";
          this.text = "";
        }
        yield {
          ...llmResponse,
          content: {
            role: llmResponse.content.role,
            parts: nonTextParts
          },
          partial: false
        };
        return;
      }
      if (sawTextPart) {
        llmResponse.partial = true;
      }
    }
    yield llmResponse;
  }
  close() {
    if (this.text || this.thoughtText) {
      const parts = [];
      if (this.thoughtText) {
        parts.push({ text: this.thoughtText, thought: true });
      }
      if (this.text) {
        parts.push({ text: this.text });
      }
      return parts;
    }
    return void 0;
  }
}
class StreamingResponseAggregator {
  constructor(isProgressiveMode = (0, import_feature_registry.isFeatureEnabled)(
    import_feature_registry.FeatureName.PROGRESSIVE_SSE_STREAMING
  )) {
    this.isProgressiveMode = isProgressiveMode;
    this.lastThoughtSignature = {};
    this.strategy = this.isProgressiveMode ? new ProgressiveStrategy() : new NonProgressiveStrategy();
  }
  async *processResponse(response) {
    var _a, _b;
    const llmResponse = (0, import_llm_response.createLlmResponse)(response);
    const parts = (_b = (_a = llmResponse.content) == null ? void 0 : _a.parts) != null ? _b : [];
    if (parts.every(isEmptyContentPart) && (llmResponse.finishReason === void 0 || llmResponse.finishReason === import_genai.FinishReason.STOP)) {
      if (llmResponse.usageMetadata) {
        this.usageMetadata = llmResponse.usageMetadata;
      }
      if (llmResponse.groundingMetadata) {
        this.groundingMetadata = llmResponse.groundingMetadata;
      }
      if (llmResponse.citationMetadata) {
        this.citationMetadata = llmResponse.citationMetadata;
      }
      if (llmResponse.finishReason) {
        this.finishReason = llmResponse.finishReason;
      }
      return;
    }
    this.response = response;
    this.usageMetadata = llmResponse.usageMetadata;
    if (llmResponse.groundingMetadata) {
      this.groundingMetadata = llmResponse.groundingMetadata;
    }
    if (llmResponse.citationMetadata) {
      this.citationMetadata = llmResponse.citationMetadata;
    }
    if (llmResponse.finishReason) {
      this.finishReason = llmResponse.finishReason;
    }
    if (llmResponse.content && llmResponse.content.parts) {
      for (const part of llmResponse.content.parts) {
        if (part.thoughtSignature) {
          this.lastThoughtSignature.value = part.thoughtSignature;
        } else if (part.functionCall && this.lastThoughtSignature.value) {
          part.thoughtSignature = this.lastThoughtSignature.value.toString();
        }
      }
    }
    yield* this.strategy.processResponse(llmResponse);
  }
  close() {
    var _a, _b, _c;
    const finalParts = this.strategy.close();
    const hasMetadata = this.usageMetadata !== void 0 || this.groundingMetadata !== void 0 || this.citationMetadata !== void 0;
    if (!finalParts && !hasMetadata) {
      return void 0;
    }
    const candidate = (_b = (_a = this.response) == null ? void 0 : _a.candidates) == null ? void 0 : _b[0];
    const finishReason = (_c = this.finishReason) != null ? _c : candidate == null ? void 0 : candidate.finishReason;
    return {
      content: {
        role: "model",
        parts: finalParts != null ? finalParts : []
      },
      groundingMetadata: this.groundingMetadata,
      citationMetadata: this.citationMetadata,
      errorCode: finishReason === import_genai.FinishReason.STOP ? void 0 : finishReason,
      errorMessage: finishReason === import_genai.FinishReason.STOP ? void 0 : candidate == null ? void 0 : candidate.finishMessage,
      usageMetadata: this.usageMetadata,
      finishReason,
      partial: false
    };
  }
}
function isEmptyContentPart(part) {
  return !part.functionCall && !part.functionResponse && !part.fileData && !part.inlineData && !part.executableCode && !part.codeExecutionResult && (!part.text || part.text === "");
}
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  StreamingResponseAggregator,
  isEmptyContentPart
});
