// @google/adk/dist/esm/tools/openapi_tool/auth/auth_helpers.js
import {createRequire as topLevelCreateRequire} from 'module';
const require = topLevelCreateRequire(import.meta.url);
/**
 * @license
 * Copyright 2026 Google LLC
 * SPDX-License-Identifier: Apache-2.0
 */
function applyCredential(url, headers, credential, authScheme) {
  if (!credential) return url;
  if (credential.apiKey) {
    let inLocation;
    let name = "key";
    if (authScheme && authScheme.type === "apiKey") {
      const apiKeyScheme = authScheme;
      inLocation = apiKeyScheme.in;
      name = apiKeyScheme.name;
    }
    if (inLocation === "header") {
      headers[name] = credential.apiKey;
    } else if (inLocation === "query") {
      const separator = url.includes("?") ? "&" : "?";
      url += `${separator}${name}=${encodeURIComponent(credential.apiKey)}`;
    } else {
      headers["Authorization"] = credential.apiKey;
    }
  } else if (credential.http && credential.http.credentials && credential.http.credentials.token) {
    headers["Authorization"] = `Bearer ${credential.http.credentials.token}`;
  }
  return url;
}
function createApiKeyScheme(name, inLocation) {
  return {
    type: "apiKey",
    name,
    in: inLocation
  };
}
function createBearerScheme() {
  return {
    type: "http",
    scheme: "bearer"
  };
}
export {
  applyCredential,
  createApiKeyScheme,
  createBearerScheme
};
