// @google/adk/dist/web/common.js
import {createRequire as topLevelCreateRequire} from 'module';
const require = topLevelCreateRequire(import.meta.url);
/**
 * @license
 * Copyright 2025 Google LLC
 * SPDX-License-Identifier: Apache-2.0
 */
import { ActiveStreamingTool } from "./agents/active_streaming_tool.js";
import { BaseAgent, isBaseAgent } from "./agents/base_agent.js";
import { Context } from "./agents/context.js";
import { functionsExportedForTestingOnly } from "./agents/functions.js";
import { InvocationContext } from "./agents/invocation_context.js";
import { LiveRequestQueue } from "./agents/live_request_queue.js";
import { LlmAgent, LlmAgent as LlmAgent2, isLlmAgent } from "./agents/llm_agent.js";
import { LoopAgent, isLoopAgent } from "./agents/loop_agent.js";
import { ParallelAgent, isParallelAgent } from "./agents/parallel_agent.js";
import { AgentTransferLlmRequestProcessor } from "./agents/processors/agent_transfer_llm_request_processor.js";
import {
  BaseLlmRequestProcessor,
  BaseLlmResponseProcessor
} from "./agents/processors/base_llm_processor.js";
import {
  CONTENT_REQUEST_PROCESSOR,
  ContentRequestProcessor
} from "./agents/processors/content_request_processor.js";
import { ContextCompactorRequestProcessor } from "./agents/processors/context_compactor_request_processor.js";
import { ReadonlyContext } from "./agents/readonly_context.js";
import { RoutedAgent, isRoutedAgent } from "./agents/routed_agent.js";
import { StreamingMode } from "./agents/run_config.js";
import { SequentialAgent, isSequentialAgent } from "./agents/sequential_agent.js";
import { InMemoryArtifactService } from "./artifacts/in_memory_artifact_service.js";
import { AuthCredentialTypes } from "./auth/auth_credential.js";
import { AuthHandler } from "./auth/auth_handler.js";
import { AUTH_PREPROCESSOR, AuthPreprocessor } from "./auth/auth_preprocessor.js";
import { AuthProviderRegistry } from "./auth/auth_provider_registry.js";
import { OAuthGrantType } from "./auth/auth_schemes.js";
import { InMemoryCredentialService } from "./auth/credential_service/in_memory_credential_service.js";
import { SessionStateCredentialService } from "./auth/credential_service/session_state_credential_service.js";
import { CredentialExchangeError } from "./auth/exchanger/base_credential_exchanger.js";
import { OAuth2CredentialExchanger } from "./auth/oauth2/oauth2_credential_exchanger.js";
import { OAuth2DiscoveryManager } from "./auth/oauth2/oauth2_discovery.js";
import { CredentialRefresherRegistry } from "./auth/refresher/credential_refresher_registry.js";
import { BaseCodeExecutor } from "./code_executors/base_code_executor.js";
import { BuiltInCodeExecutor } from "./code_executors/built_in_code_executor.js";
import {
  CodeExecutionLanguage,
  FileContentEncoding
} from "./code_executors/code_execution_utils.js";
import { LlmSummarizer } from "./context/summarizers/llm_summarizer.js";
import { TokenBasedContextCompactor } from "./context/token_based_context_compactor.js";
import { TruncatingContextCompactor } from "./context/truncating_context_compactor.js";
import { isCompactedEvent } from "./events/compacted_event.js";
import {
  createEvent,
  getFunctionCalls,
  getFunctionResponses,
  hasTrailingCodeExecutionResult,
  isFinalResponse,
  stringifyContent
} from "./events/event.js";
import { createEventActions } from "./events/event_actions.js";
import { EventType, toStructuredEvents } from "./events/structured_events.js";
import {
  BaseExampleProvider,
  isBaseExampleProvider
} from "./examples/base_example_provider.js";
import { InMemoryMemoryService } from "./memory/in_memory_memory_service.js";
import { VertexAiMemoryBankService } from "./memory/vertex_ai_memory_bank_service.js";
import { ApigeeLlm } from "./models/apigee_llm.js";
import { BaseLlm, isBaseLlm } from "./models/base_llm.js";
import { Gemini, geminiInitParams } from "./models/google_llm.js";
import { LLMRegistry } from "./models/registry.js";
import { RoutedLlm } from "./models/routed_llm.js";
import { BasePlugin, ContextCompactionTrigger } from "./plugins/base_plugin.js";
import { LoggingPlugin } from "./plugins/logging_plugin.js";
import { PluginManager } from "./plugins/plugin_manager.js";
import {
  InMemoryPolicyEngine,
  PolicyOutcome,
  REQUEST_CONFIRMATION_FUNCTION_CALL_NAME,
  SecurityPlugin,
  getAskUserConfirmationFunctionCalls
} from "./plugins/security_plugin.js";
import { InMemoryRunner } from "./runner/in_memory_runner.js";
import { Runner, isRunner } from "./runner/runner.js";
import { BaseSessionService } from "./sessions/base_session_service.js";
import { InMemorySessionService } from "./sessions/in_memory_session_service.js";
import { createSession } from "./sessions/session.js";
import { State } from "./sessions/state.js";
import { AgentTool, isAgentTool } from "./tools/agent_tool.js";
import { BaseTool, isBaseTool } from "./tools/base_tool.js";
import { BaseToolset, isBaseToolset } from "./tools/base_toolset.js";
import { EXIT_LOOP, ExitLoopTool } from "./tools/exit_loop_tool.js";
import { FunctionTool, isFunctionTool } from "./tools/function_tool.js";
import {
  GOOGLE_MAPS_GROUNDING,
  GoogleMapsGroundingTool
} from "./tools/google_maps_grounding_tool.js";
import { GOOGLE_SEARCH, GoogleSearchTool } from "./tools/google_search_tool.js";
import {
  LOAD_ARTIFACTS,
  LoadArtifactsTool
} from "./tools/load_artifacts_tool.js";
import { LOAD_MEMORY, LoadMemoryTool } from "./tools/load_memory_tool.js";
import { LongRunningFunctionTool } from "./tools/long_running_tool.js";
import {
  PRELOAD_MEMORY,
  PreloadMemoryTool
} from "./tools/preload_memory_tool.js";
import { ToolConfirmation } from "./tools/tool_confirmation.js";
import { URL_CONTEXT, UrlContextTool } from "./tools/url_context_tool.js";
import { VertexAiSearchTool } from "./tools/vertex_ai_search_tool.js";
import { VertexRagRetrievalTool } from "./tools/vertex_rag_retrieval_tool.js";
import { LogLevel, getLogger, setLogLevel, setLogger } from "./utils/logger.js";
import { isGemini2OrAbove, isGemini3xFlashLive } from "./utils/model_name.js";
import { zodObjectToSchema } from "./utils/simple_zod_to_json.js";
import { GoogleLLMVariant } from "./utils/variant_utils.js";
import { version } from "./version.js";
import { GCPSkillRegistry } from "./skills/gcp_skill_registry.js";
import {
  loadAllSkillsInDir,
  loadSkillFromDir,
  loadSkillFromZipBuffer,
  validateSkillDir
} from "./skills/loader.js";
import { ListSkillsTool } from "./tools/skill/list_skills_tool.js";
import { LoadSkillResourceTool } from "./tools/skill/load_skill_resource_tool.js";
import { LoadSkillTool } from "./tools/skill/load_skill_tool.js";
import { SearchSkillsTool } from "./tools/skill/search_skills_tool.js";
import { SkillToolset } from "./tools/skill/skill_toolset.js";
import { OpenApiSpecParser } from "./tools/openapi_tool/openapi_spec_parser/openapi_spec_parser.js";
import { OperationParser } from "./tools/openapi_tool/openapi_spec_parser/operation_parser.js";
import { ToolAuthHandler } from "./tools/openapi_tool/openapi_spec_parser/tool_auth_handler.js";
import { OpenAPIToolset } from "./tools/openapi_tool/openapi_toolset.js";
import {
  RestApiTool,
  createRestApiTool
} from "./tools/openapi_tool/rest_api_tool.js";
export * from "./artifacts/base_artifact_service.js";
export * from "./features/feature_registry.js";
export * from "./memory/base_memory_service.js";
export * from "./sessions/base_session_service.js";
export * from "./tools/base_tool.js";
export {
  AUTH_PREPROCESSOR,
  ActiveStreamingTool,
  LlmAgent as Agent,
  AgentTool,
  AgentTransferLlmRequestProcessor,
  ApigeeLlm,
  AuthCredentialTypes,
  AuthHandler,
  AuthPreprocessor,
  AuthProviderRegistry,
  BaseAgent,
  BaseCodeExecutor,
  BaseExampleProvider,
  BaseLlm,
  BaseLlmRequestProcessor,
  BaseLlmResponseProcessor,
  BasePlugin,
  BaseSessionService,
  BaseTool,
  BaseToolset,
  BuiltInCodeExecutor,
  CONTENT_REQUEST_PROCESSOR,
  CodeExecutionLanguage,
  ContentRequestProcessor,
  Context,
  ContextCompactionTrigger,
  ContextCompactorRequestProcessor,
  CredentialExchangeError,
  CredentialRefresherRegistry,
  EXIT_LOOP,
  EventType,
  ExitLoopTool,
  FileContentEncoding,
  FunctionTool,
  GCPSkillRegistry,
  GOOGLE_MAPS_GROUNDING,
  GOOGLE_SEARCH,
  Gemini,
  GoogleLLMVariant,
  GoogleMapsGroundingTool,
  GoogleSearchTool,
  InMemoryArtifactService,
  InMemoryCredentialService,
  InMemoryMemoryService,
  InMemoryPolicyEngine,
  InMemoryRunner,
  InMemorySessionService,
  InvocationContext,
  LLMRegistry,
  LOAD_ARTIFACTS,
  LOAD_MEMORY,
  ListSkillsTool,
  LiveRequestQueue,
  LlmAgent2 as LlmAgent,
  LlmSummarizer,
  LoadArtifactsTool,
  LoadMemoryTool,
  LoadSkillResourceTool,
  LoadSkillTool,
  LogLevel,
  LoggingPlugin,
  LongRunningFunctionTool,
  LoopAgent,
  OAuth2CredentialExchanger,
  OAuth2DiscoveryManager,
  OAuthGrantType,
  OpenAPIToolset,
  OpenApiSpecParser,
  OperationParser,
  PRELOAD_MEMORY,
  ParallelAgent,
  PluginManager,
  PolicyOutcome,
  PreloadMemoryTool,
  REQUEST_CONFIRMATION_FUNCTION_CALL_NAME,
  ReadonlyContext,
  RestApiTool,
  RoutedAgent,
  RoutedLlm,
  Runner,
  SearchSkillsTool,
  SecurityPlugin,
  SequentialAgent,
  SessionStateCredentialService,
  SkillToolset,
  State,
  StreamingMode,
  TokenBasedContextCompactor,
  ToolAuthHandler,
  ToolConfirmation,
  TruncatingContextCompactor,
  URL_CONTEXT,
  UrlContextTool,
  VertexAiMemoryBankService,
  VertexAiSearchTool,
  VertexRagRetrievalTool,
  createEvent,
  createEventActions,
  createRestApiTool,
  createSession,
  functionsExportedForTestingOnly,
  geminiInitParams,
  getAskUserConfirmationFunctionCalls,
  getFunctionCalls,
  getFunctionResponses,
  getLogger,
  hasTrailingCodeExecutionResult,
  isAgentTool,
  isBaseAgent,
  isBaseExampleProvider,
  isBaseLlm,
  isBaseTool,
  isBaseToolset,
  isCompactedEvent,
  isFinalResponse,
  isFunctionTool,
  isGemini2OrAbove,
  isGemini3xFlashLive,
  isLlmAgent,
  isLoopAgent,
  isParallelAgent,
  isRoutedAgent,
  isRunner,
  isSequentialAgent,
  loadAllSkillsInDir,
  loadSkillFromDir,
  loadSkillFromZipBuffer,
  setLogLevel,
  setLogger,
  stringifyContent,
  toStructuredEvents,
  validateSkillDir,
  version,
  zodObjectToSchema
};
