// @google/adk/dist/cjs/tools/openapi_tool/auth/auth_helpers.js
/**
  * @license
  * Copyright 2026 Google LLC
  * SPDX-License-Identifier: Apache-2.0
  */

"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var auth_helpers_exports = {};
__export(auth_helpers_exports, {
  applyCredential: () => applyCredential,
  createApiKeyScheme: () => createApiKeyScheme,
  createBearerScheme: () => createBearerScheme
});
module.exports = __toCommonJS(auth_helpers_exports);
/**
 * @license
 * Copyright 2026 Google LLC
 * SPDX-License-Identifier: Apache-2.0
 */
function applyCredential(url, headers, credential, authScheme) {
  if (!credential) return url;
  if (credential.apiKey) {
    let inLocation;
    let name = "key";
    if (authScheme && authScheme.type === "apiKey") {
      const apiKeyScheme = authScheme;
      inLocation = apiKeyScheme.in;
      name = apiKeyScheme.name;
    }
    if (inLocation === "header") {
      headers[name] = credential.apiKey;
    } else if (inLocation === "query") {
      const separator = url.includes("?") ? "&" : "?";
      url += `${separator}${name}=${encodeURIComponent(credential.apiKey)}`;
    } else {
      headers["Authorization"] = credential.apiKey;
    }
  } else if (credential.http && credential.http.credentials && credential.http.credentials.token) {
    headers["Authorization"] = `Bearer ${credential.http.credentials.token}`;
  }
  return url;
}
function createApiKeyScheme(name, inLocation) {
  return {
    type: "apiKey",
    name,
    in: inLocation
  };
}
function createBearerScheme() {
  return {
    type: "http",
    scheme: "bearer"
  };
}
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  applyCredential,
  createApiKeyScheme,
  createBearerScheme
});
