// @google/adk/dist/web/utils/async_queue.js
import {createRequire as topLevelCreateRequire} from 'module';
const require = topLevelCreateRequire(import.meta.url);
/**
 * @license
 * Copyright 2026 Google LLC
 * SPDX-License-Identifier: Apache-2.0
 */
class AsyncQueue {
  constructor() {
    this.queue = [];
    this.resolvers = [];
    this.closed = false;
  }
  push(value) {
    if (this.closed) return;
    if (this.resolvers.length > 0) {
      const { resolve } = this.resolvers.shift();
      resolve({ value, done: false });
    } else {
      this.queue.push(value);
    }
  }
  error(err) {
    this.errorVal = err;
    while (this.resolvers.length > 0) {
      const { reject } = this.resolvers.shift();
      reject(err);
    }
  }
  close() {
    this.closed = true;
    while (this.resolvers.length > 0) {
      const { resolve } = this.resolvers.shift();
      resolve({ value: void 0, done: true });
    }
  }
  [Symbol.asyncIterator]() {
    return {
      next: () => {
        if (this.errorVal) {
          const err = this.errorVal;
          this.errorVal = void 0;
          return Promise.reject(err);
        }
        if (this.queue.length > 0) {
          return Promise.resolve({ value: this.queue.shift(), done: false });
        }
        if (this.closed) {
          return Promise.resolve({ value: void 0, done: true });
        }
        return new Promise((resolve, reject) => {
          this.resolvers.push({ resolve, reject });
        });
      }
    };
  }
}
export {
  AsyncQueue
};
