// @google/adk/dist/web/utils/streaming_utils.js
import {createRequire as topLevelCreateRequire} from 'module';
const require = topLevelCreateRequire(import.meta.url);
var __defProp = Object.defineProperty;
var __defProps = Object.defineProperties;
var __getOwnPropDescs = Object.getOwnPropertyDescriptors;
var __getOwnPropSymbols = Object.getOwnPropertySymbols;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __propIsEnum = Object.prototype.propertyIsEnumerable;
var __knownSymbol = (name, symbol) => (symbol = Symbol[name]) ? symbol : Symbol.for("Symbol." + name);
var __typeError = (msg) => {
  throw TypeError(msg);
};
var __defNormalProp = (obj, key, value) => key in obj ? __defProp(obj, key, { enumerable: true, configurable: true, writable: true, value }) : obj[key] = value;
var __spreadValues = (a, b) => {
  for (var prop in b || (b = {}))
    if (__hasOwnProp.call(b, prop))
      __defNormalProp(a, prop, b[prop]);
  if (__getOwnPropSymbols)
    for (var prop of __getOwnPropSymbols(b)) {
      if (__propIsEnum.call(b, prop))
        __defNormalProp(a, prop, b[prop]);
    }
  return a;
};
var __spreadProps = (a, b) => __defProps(a, __getOwnPropDescs(b));
var __await = function(promise, isYieldStar) {
  this[0] = promise;
  this[1] = isYieldStar;
};
var __asyncGenerator = (__this, __arguments, generator) => {
  var resume = (k, v, yes, no) => {
    try {
      var x = generator[k](v), isAwait = (v = x.value) instanceof __await, done = x.done;
      Promise.resolve(isAwait ? v[0] : v).then((y) => isAwait ? resume(k === "return" ? k : "next", v[1] ? { done: y.done, value: y.value } : y, yes, no) : yes({ value: y, done })).catch((e) => resume("throw", e, yes, no));
    } catch (e) {
      no(e);
    }
  }, method = (k) => it[k] = (x) => new Promise((yes, no) => resume(k, x, yes, no)), it = {};
  return generator = generator.apply(__this, __arguments), it[__knownSymbol("asyncIterator")] = () => it, method("next"), method("throw"), method("return"), it;
};
var __yieldStar = (value) => {
  var obj = value[__knownSymbol("asyncIterator")], isAwait = false, method, it = {};
  if (obj == null) {
    obj = value[__knownSymbol("iterator")]();
    method = (k) => it[k] = (x) => obj[k](x);
  } else {
    obj = obj.call(value);
    method = (k) => it[k] = (v) => {
      if (isAwait) {
        isAwait = false;
        if (k === "throw") throw v;
        return v;
      }
      isAwait = true;
      return {
        done: false,
        value: new __await(new Promise((resolve) => {
          var x = obj[k](v);
          if (!(x instanceof Object)) __typeError("Object expected");
          resolve(x);
        }), 1)
      };
    };
  }
  return it[__knownSymbol("iterator")] = () => it, method("next"), "throw" in obj ? method("throw") : it.throw = (x) => {
    throw x;
  }, "return" in obj && method("return"), it;
};
/**
 * @license
 * Copyright 2026 Google LLC
 * SPDX-License-Identifier: Apache-2.0
 */
import {
  FinishReason
} from "@google/genai";
import { JSONPath } from "jsonpath-plus";
import { generateClientFunctionCallId } from "../agents/functions.js";
import { FeatureName, isFeatureEnabled } from "../features/feature_registry.js";
import { createLlmResponse } from "../models/llm_response.js";
const UNSAFE_PROPERTY_KEYS = /* @__PURE__ */ new Set(["__proto__", "constructor", "prototype"]);
class ProgressiveStrategy {
  constructor() {
    this.partsSequence = [];
    this.currentTextBuffer = "";
    this.currentFcArgs = {};
  }
  flushTextBufferToSequence() {
    if (!this.currentTextBuffer) {
      return;
    }
    if (this.currentTextIsThought) {
      this.partsSequence.push({
        text: this.currentTextBuffer,
        thought: true
      });
    } else {
      this.partsSequence.push({
        text: this.currentTextBuffer
      });
    }
    this.currentTextBuffer = "";
    this.currentTextIsThought = void 0;
  }
  getValueFromPartialArg(partialArg, jsonPath) {
    let value = null;
    let hasValue = false;
    const stringValue = partialArg.stringValue;
    const numberValue = partialArg.numberValue;
    const boolValue = partialArg.boolValue;
    const nullValue = partialArg.nullValue;
    if (stringValue !== void 0) {
      const stringChunk = stringValue;
      hasValue = true;
      const pathParts = JSONPath.toPathArray(jsonPath).filter(
        (p) => p !== "$" && p !== "$["
      );
      let existingValue = this.currentFcArgs;
      for (const part of pathParts) {
        if (existingValue && typeof existingValue === "object" && part in existingValue) {
          existingValue = existingValue[part];
        } else {
          existingValue = void 0;
          break;
        }
      }
      if (typeof existingValue === "string") {
        value = existingValue + stringChunk;
      } else {
        value = stringChunk;
      }
    } else if (numberValue !== void 0) {
      value = numberValue;
      hasValue = true;
    } else if (boolValue !== void 0) {
      value = boolValue;
      hasValue = true;
    } else if (nullValue !== void 0) {
      value = null;
      hasValue = true;
    }
    return [value, hasValue];
  }
  setValueByJsonPath(jsonPath, value) {
    const pathParts = JSONPath.toPathArray(jsonPath).filter(
      (p) => p !== "$" && p !== "$["
    );
    if (pathParts.some((part) => UNSAFE_PROPERTY_KEYS.has(String(part)))) {
      return;
    }
    let current = this.currentFcArgs;
    for (let i = 0; i < pathParts.length - 1; i++) {
      const part = pathParts[i];
      if (!Object.prototype.hasOwnProperty.call(current, part)) {
        current[part] = {};
      }
      current = current[part];
    }
    if (pathParts.length > 0) {
      current[pathParts[pathParts.length - 1]] = value;
    }
  }
  flushFunctionCallToSequence() {
    var _a;
    if (this.currentFcName) {
      const fcPart = {
        functionCall: {
          name: this.currentFcName,
          args: JSON.parse(JSON.stringify(this.currentFcArgs)),
          id: (_a = this.currentFcId) != null ? _a : generateClientFunctionCallId()
        }
      };
      if (this.currentThoughtSignature) {
        fcPart.thoughtSignature = this.currentThoughtSignature.toString();
      }
      this.partsSequence.push(fcPart);
      this.currentFcName = void 0;
      this.currentFcArgs = {};
      this.currentFcId = void 0;
      this.currentThoughtSignature = void 0;
    }
  }
  processStreamingFunctionCall(fc) {
    if (fc.name) {
      this.currentFcName = fc.name;
    }
    if (fc.id) {
      this.currentFcId = fc.id;
    }
    for (const partialArg of fc.partialArgs || []) {
      const jsonPath = partialArg.jsonPath;
      if (!jsonPath) {
        continue;
      }
      const [value, hasValue] = this.getValueFromPartialArg(
        partialArg,
        jsonPath
      );
      if (hasValue) {
        this.setValueByJsonPath(jsonPath, value);
      }
    }
    if (!fc.willContinue) {
      this.flushTextBufferToSequence();
      this.flushFunctionCallToSequence();
    }
  }
  processFunctionCallPart(part) {
    const fc = part.functionCall;
    if (!fc) {
      return;
    }
    if (fc.partialArgs || fc.willContinue) {
      if (!fc.id && !this.currentFcId) {
        fc.id = generateClientFunctionCallId();
      }
      if (part.thoughtSignature && !this.currentThoughtSignature) {
        this.currentThoughtSignature = part.thoughtSignature;
      }
      this.processStreamingFunctionCall(fc);
    } else {
      if (fc.name) {
        if (!fc.id) {
          fc.id = generateClientFunctionCallId();
        }
        this.flushTextBufferToSequence();
        this.partsSequence.push(part);
      }
    }
  }
  processResponse(llmResponse) {
    return __asyncGenerator(this, null, function* () {
      var _a;
      if (llmResponse.content && llmResponse.content.parts) {
        for (const part of llmResponse.content.parts) {
          if (part.text) {
            const isThought = (_a = part.thought) != null ? _a : false;
            if (this.currentTextBuffer && isThought !== this.currentTextIsThought) {
              this.flushTextBufferToSequence();
            }
            if (!this.currentTextBuffer) {
              this.currentTextIsThought = isThought;
            }
            this.currentTextBuffer += part.text;
          } else if (part.functionCall) {
            this.processFunctionCallPart(part);
          } else {
            this.flushTextBufferToSequence();
            this.partsSequence.push(part);
          }
        }
      }
      llmResponse.partial = true;
      yield llmResponse;
    });
  }
  close() {
    this.flushTextBufferToSequence();
    this.flushFunctionCallToSequence();
    const finalParts = this.partsSequence;
    if (finalParts.length === 0) {
      return void 0;
    }
    return finalParts;
  }
}
class NonProgressiveStrategy {
  constructor() {
    this.text = "";
    this.thoughtText = "";
  }
  processResponse(llmResponse) {
    return __asyncGenerator(this, null, function* () {
      var _a;
      if ((_a = llmResponse.content) == null ? void 0 : _a.parts) {
        const nonTextParts = [];
        let sawTextPart = false;
        for (const part of llmResponse.content.parts) {
          if (typeof part.text === "string") {
            sawTextPart = true;
            if (part.thought) {
              this.thoughtText += part.text;
            } else {
              this.text += part.text;
            }
            continue;
          }
          if (part.functionCall && !part.functionCall.id) {
            part.functionCall.id = generateClientFunctionCallId();
          }
          nonTextParts.push(part);
        }
        if (nonTextParts.length > 0) {
          if (this.thoughtText || this.text) {
            const parts = [];
            if (this.thoughtText) {
              parts.push({ text: this.thoughtText, thought: true });
            }
            if (this.text) {
              parts.push({ text: this.text });
            }
            yield {
              content: {
                role: "model",
                parts
              },
              usageMetadata: llmResponse.usageMetadata,
              partial: false
            };
            this.thoughtText = "";
            this.text = "";
          }
          yield __spreadProps(__spreadValues({}, llmResponse), {
            content: {
              role: llmResponse.content.role,
              parts: nonTextParts
            },
            partial: false
          });
          return;
        }
        if (sawTextPart) {
          llmResponse.partial = true;
        }
      }
      yield llmResponse;
    });
  }
  close() {
    if (this.text || this.thoughtText) {
      const parts = [];
      if (this.thoughtText) {
        parts.push({ text: this.thoughtText, thought: true });
      }
      if (this.text) {
        parts.push({ text: this.text });
      }
      return parts;
    }
    return void 0;
  }
}
class StreamingResponseAggregator {
  constructor(isProgressiveMode = isFeatureEnabled(
    FeatureName.PROGRESSIVE_SSE_STREAMING
  )) {
    this.isProgressiveMode = isProgressiveMode;
    this.lastThoughtSignature = {};
    this.strategy = this.isProgressiveMode ? new ProgressiveStrategy() : new NonProgressiveStrategy();
  }
  processResponse(response) {
    return __asyncGenerator(this, null, function* () {
      var _a, _b;
      const llmResponse = createLlmResponse(response);
      const parts = (_b = (_a = llmResponse.content) == null ? void 0 : _a.parts) != null ? _b : [];
      if (parts.every(isEmptyContentPart) && (llmResponse.finishReason === void 0 || llmResponse.finishReason === FinishReason.STOP)) {
        if (llmResponse.usageMetadata) {
          this.usageMetadata = llmResponse.usageMetadata;
        }
        if (llmResponse.groundingMetadata) {
          this.groundingMetadata = llmResponse.groundingMetadata;
        }
        if (llmResponse.citationMetadata) {
          this.citationMetadata = llmResponse.citationMetadata;
        }
        if (llmResponse.finishReason) {
          this.finishReason = llmResponse.finishReason;
        }
        return;
      }
      this.response = response;
      this.usageMetadata = llmResponse.usageMetadata;
      if (llmResponse.groundingMetadata) {
        this.groundingMetadata = llmResponse.groundingMetadata;
      }
      if (llmResponse.citationMetadata) {
        this.citationMetadata = llmResponse.citationMetadata;
      }
      if (llmResponse.finishReason) {
        this.finishReason = llmResponse.finishReason;
      }
      if (llmResponse.content && llmResponse.content.parts) {
        for (const part of llmResponse.content.parts) {
          if (part.thoughtSignature) {
            this.lastThoughtSignature.value = part.thoughtSignature;
          } else if (part.functionCall && this.lastThoughtSignature.value) {
            part.thoughtSignature = this.lastThoughtSignature.value.toString();
          }
        }
      }
      yield* __yieldStar(this.strategy.processResponse(llmResponse));
    });
  }
  close() {
    var _a, _b, _c;
    const finalParts = this.strategy.close();
    const hasMetadata = this.usageMetadata !== void 0 || this.groundingMetadata !== void 0 || this.citationMetadata !== void 0;
    if (!finalParts && !hasMetadata) {
      return void 0;
    }
    const candidate = (_b = (_a = this.response) == null ? void 0 : _a.candidates) == null ? void 0 : _b[0];
    const finishReason = (_c = this.finishReason) != null ? _c : candidate == null ? void 0 : candidate.finishReason;
    return {
      content: {
        role: "model",
        parts: finalParts != null ? finalParts : []
      },
      groundingMetadata: this.groundingMetadata,
      citationMetadata: this.citationMetadata,
      errorCode: finishReason === FinishReason.STOP ? void 0 : finishReason,
      errorMessage: finishReason === FinishReason.STOP ? void 0 : candidate == null ? void 0 : candidate.finishMessage,
      usageMetadata: this.usageMetadata,
      finishReason,
      partial: false
    };
  }
}
function isEmptyContentPart(part) {
  return !part.functionCall && !part.functionResponse && !part.fileData && !part.inlineData && !part.executableCode && !part.codeExecutionResult && (!part.text || part.text === "");
}
export {
  StreamingResponseAggregator,
  isEmptyContentPart
};
